{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping by sequencing pipeline\n",
    "## Parental strain analysis - Hawaiian tutorial\n",
    "* [Amhed Missael Vargas Velazquez](https://www.researchgate.net/profile/Amhed-Vargas-Velazquez)\n",
    "* Post-doctoral fellow, [SGB lab](https://syngenbio.kaust.edu.sa/), [KAUST](https://www.kaust.edu.sa/en)\n",
    "\n",
    "## Description\n",
    "This jupyter notebook will guide you through the computation used to identify genomic variants in the highly polymorphic *C. elegans* strain CB4856 which was isolated in Hawaii (hence its alias as *Hawaiian*). This step is essential in the [mapping by sequencing](https://www.genetics.org/content/204/2/451) pipeline only if the mapping strain, *a.k.a*. the parental strain, had not been analized before, *i.e.*, you don't have a .vcf file containing background genomic variants. \n",
    "\n",
    "## Software requirements\n",
    "The core instance running this script is Python2. However, the analysis are performed by multiple programs (handled by `system calls`) which have to be installed or, for portability convenience, be present in a folder (check WGS-Software_configuration notebook for further details). Note that software in **bold** must be installed and compiled in the executing computer because they cannot be used as stand-alone versions.\n",
    "\n",
    "* [**samtools**](http://www.htslib.org/)\n",
    "* [**bwa**](https://github.com/lh3/bwa)\n",
    "* [GATK](https://gatk.broadinstitute.org/hc/en-us) [3.7+](https://console.cloud.google.com/storage/browser/gatk-software/package-archive/gatk/); note that current version of GATK (4.0+) uses different nomenclature and have different tools which makes it not compatible with this pipeline.\n",
    "* [picard](https://broadinstitute.github.io/picard/)\n",
    "* [snpEff](https://pcingola.github.io/SnpEff/)\n",
    "* [fastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)\n",
    "* [SRAToolkit](https://github.com/ncbi/sra-tools)\n",
    "* wget\n",
    "* zcat\n",
    "* [**R**](https://cran.r-project.org/)\n",
    "* Python2\n",
    "* Perl\n",
    "\n",
    "### R libraries\n",
    "* ggplot2\n",
    "\n",
    "### Prerequisites\n",
    "Before running any code, make sure to select which folder where the analysis will be performed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load libraries\n",
    "#os to move within directories\n",
    "import os\n",
    "##Set main working directory\n",
    "#Same location as script\n",
    "#path = '/home/jupyter-user/'\n",
    "path = os.getcwd()\n",
    "##Move to path\n",
    "os.chdir(path)\n",
    "#Show current directory\n",
    "print('Main directory:\\n' + os.getcwd())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### A reference genome\n",
    "In order to run this pipeline, a *C. elegans* reference genome (WS235/ce11) is needed; you can download the most recent ensemble version (ce11) by running the following code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘data’: File exists\r\n"
     ]
    }
   ],
   "source": [
    "##Create directory to place data\n",
    "!mkdir -p {path}/data\n",
    "##Move to that directory\n",
    "os.chdir(path+'/data')\n",
    "#Download the C. elegans genome from ensemble release 99 (same version as SnpEff)\n",
    "!wget -q ftp://ftp.ensembl.org/pub/release-99/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.dna_sm.toplevel.fa.gz\n",
    "#Uncompress and rename\n",
    "!zcat Caenorhabditis_elegans.WBcel235.dna_sm.toplevel.fa.gz > Caenorhabditis_elegans.WBcel235.99.softmasked.fa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then it need to be indexed by samtools:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools faidx Caenorhabditis_elegans.WBcel235.99.softmasked.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And bwa:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.59 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=200572802, availableWord=26113028\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 43074434 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 79575970 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 112014578 characters processed.\n",
      "[BWTIncConstructFromPacked] 40 iterations done. 140842098 characters processed.\n",
      "[BWTIncConstructFromPacked] 50 iterations done. 166460034 characters processed.\n",
      "[BWTIncConstructFromPacked] 60 iterations done. 189225266 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 66 iterations.\n",
      "[bwa_index] 42.12 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.45 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.34 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 21.76 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index Caenorhabditis_elegans.WBcel235.99.softmasked.fa\n",
      "[main] Real time: 67.565 sec; CPU: 65.264 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa index Caenorhabditis_elegans.WBcel235.99.softmasked.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stand alone software\n",
    "Most of the programs used within this pipeline have \"stand-alone\" versions that allow users to run their analysis on any computer they want. However, first you have to make sure to have those programs. Particularly, make sure to have a directory containing the following ones: \n",
    "\n",
    "- GATK (GenomeAnalysisTK.jar)\n",
    "- Picard (picard.jar)\n",
    "- SnpEff (folder with both snpEff.jar and SnpSift.jar, and another folder with its database; more below)\n",
    "\n",
    "For your convenience, we distribute a folder already containing these programs. Just make sure to set properly the path to them, e.g.:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Path to software folder\n",
    "softpath = '/home/WGS_pipeline/Software/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the case that a new version of a program is needed or the software configuration has changed, please check the jupyter notebook `WGS-Software_configuration`. This notebook indicates where and how to intall all the programs needed for the pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "Whole Genome Sequencing (WGS) has revolutionized how we identify genes that act in the development of a phenotype. For a short historical review check my [lab presentation](https://docs.google.com/presentation/d/1pAx1WlkFoawavdneXdca28Gdr7n6Yfc4EdoAf7yVIqc/edit?usp=sharing) and for a better understanding read the Wormbook chapter on [NGS methods to map mutations in *C. elegans*](https://www.genetics.org/content/204/2/451).\n",
    "\n",
    "This jupyter notebook will guide you through the computation used to identify genomic variants in the highly polymorphic C. elegans strain CB4856 which was isolated in Hawaii (hence its alias as Hawaiian). This is the first notebook of a series of tutorials"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hawaiian strain\n",
    "\n",
    "#### Reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets download fastq reads from SRA cloud hosted in Amazon Web Services. We would use `wget` to download the SRA file and the SRAToolkit to dump them into fastq files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 19794919 spots for SRR2003569\n",
      "Written 19794919 spots for SRR2003569\n",
      "Caenorhabditis_elegans.WBcel235.99.softmasked.fa\n",
      "Caenorhabditis_elegans.WBcel235.99.softmasked.fa.amb\n",
      "Caenorhabditis_elegans.WBcel235.99.softmasked.fa.ann\n",
      "Caenorhabditis_elegans.WBcel235.99.softmasked.fa.bwt\n",
      "Caenorhabditis_elegans.WBcel235.99.softmasked.fa.fai\n",
      "Caenorhabditis_elegans.WBcel235.99.softmasked.fa.pac\n",
      "Caenorhabditis_elegans.WBcel235.99.softmasked.fa.sa\n",
      "Caenorhabditis_elegans.WBcel235.dna_sm.toplevel.fa.gz\n",
      "Caenorhabditis_elegans.WBcel235.dna_sm.toplevel.fa.gz.1\n",
      "SRR2003569\n",
      "SRR2003569_1.fastq\n",
      "SRR2003569_2.fastq\n"
     ]
    }
   ],
   "source": [
    "## SRR2003569: https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR2003569\n",
    "!wget -q https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR2003569/SRR2003569\n",
    "\n",
    "##Dump files into fastq files\n",
    "!fastq-dump --split-files SRR2003569\n",
    "\n",
    "##Verify if files were properly downloaded\n",
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets go out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Move to that directory\n",
    "os.chdir(path)\n",
    "##Create directory to place data\n",
    "!mkdir -p {path}/analysis\n",
    "##Move to that directory\n",
    "os.chdir( path + '/analysis' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets now define parameters for pipeline:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Inputs\n",
    "SampleName = 'CB4856'\n",
    "FastqF1 = (path + '/data/SRR2003569_1.fastq') \n",
    "FastqF2 = (path + '/data/SRR2003569_2.fastq')\n",
    "ReferenceGenome = (path + '/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa')\n",
    "SnpEffGen = 'WBcel235.99'\n",
    "\n",
    "#Minimal quality for bam\n",
    "minMaqQforBam = 1\n",
    "#Minimum quality for vcf filtering\n",
    "minMaqQforVcf = 10\n",
    "#Minimum basequality for variant calling\n",
    "minBaseQforVcf = 10\n",
    "#\n",
    "minVarCall = 10\n",
    "diffStep = 10\n",
    "minDel = 2\n",
    "maxSample= 2\n",
    "\n",
    "#Number of threats\n",
    "Ncpu=20\n",
    "#Ram for java\n",
    "ramG=40"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FastQC interpretation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Create directory to place data\n",
    "!mkdir -p FastQC\n",
    "##Move to that directory\n",
    "os.chdir(path + '/analysis/FastQC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started analysis of SRR2003569_1.fastq\n",
      "Started analysis of SRR2003569_2.fastq\n",
      "Approx 5% complete for SRR2003569_2.fastq\n",
      "Approx 5% complete for SRR2003569_1.fastq\n",
      "Approx 10% complete for SRR2003569_2.fastq\n",
      "Approx 10% complete for SRR2003569_1.fastq\n",
      "Approx 15% complete for SRR2003569_2.fastq\n",
      "Approx 15% complete for SRR2003569_1.fastq\n",
      "Approx 20% complete for SRR2003569_2.fastq\n",
      "Approx 20% complete for SRR2003569_1.fastq\n",
      "Approx 25% complete for SRR2003569_2.fastq\n",
      "Approx 25% complete for SRR2003569_1.fastq\n",
      "Approx 30% complete for SRR2003569_2.fastq\n",
      "Approx 35% complete for SRR2003569_2.fastq\n",
      "Approx 30% complete for SRR2003569_1.fastq\n",
      "Approx 40% complete for SRR2003569_2.fastq\n",
      "Approx 35% complete for SRR2003569_1.fastq\n",
      "Approx 45% complete for SRR2003569_2.fastq\n",
      "Approx 40% complete for SRR2003569_1.fastq\n",
      "Approx 50% complete for SRR2003569_2.fastq\n",
      "Approx 45% complete for SRR2003569_1.fastq\n",
      "Approx 55% complete for SRR2003569_2.fastq\n",
      "Approx 50% complete for SRR2003569_1.fastq\n",
      "Approx 60% complete for SRR2003569_2.fastq\n",
      "Approx 65% complete for SRR2003569_2.fastq\n",
      "Approx 55% complete for SRR2003569_1.fastq\n",
      "Approx 70% complete for SRR2003569_2.fastq\n",
      "Approx 60% complete for SRR2003569_1.fastq\n",
      "Approx 75% complete for SRR2003569_2.fastq\n",
      "Approx 65% complete for SRR2003569_1.fastq\n",
      "Approx 80% complete for SRR2003569_2.fastq\n",
      "Approx 70% complete for SRR2003569_1.fastq\n",
      "Approx 85% complete for SRR2003569_2.fastq\n",
      "Approx 75% complete for SRR2003569_1.fastq\n",
      "Approx 90% complete for SRR2003569_2.fastq\n",
      "Approx 80% complete for SRR2003569_1.fastq\n",
      "Approx 95% complete for SRR2003569_2.fastq\n",
      "Approx 85% complete for SRR2003569_1.fastq\n",
      "Analysis complete for SRR2003569_2.fastq\n",
      "Approx 90% complete for SRR2003569_1.fastq\n",
      "Approx 95% complete for SRR2003569_1.fastq\n",
      "Analysis complete for SRR2003569_1.fastq\n"
     ]
    }
   ],
   "source": [
    "##Perform fastQC analysis\n",
    "!fastqc -t {Ncpu} {FastqF1} {FastqF2} -o ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Basic assesment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "2 archives were successfully processed.\n",
      "SRR2003569_1\n",
      "PASS\tBasic Statistics\tSRR2003569_1.fastq\n",
      "PASS\tPer base sequence quality\tSRR2003569_1.fastq\n",
      "WARN\tPer tile sequence quality\tSRR2003569_1.fastq\n",
      "PASS\tPer sequence quality scores\tSRR2003569_1.fastq\n",
      "FAIL\tPer base sequence content\tSRR2003569_1.fastq\n",
      "PASS\tPer sequence GC content\tSRR2003569_1.fastq\n",
      "PASS\tPer base N content\tSRR2003569_1.fastq\n",
      "PASS\tSequence Length Distribution\tSRR2003569_1.fastq\n",
      "WARN\tSequence Duplication Levels\tSRR2003569_1.fastq\n",
      "PASS\tOverrepresented sequences\tSRR2003569_1.fastq\n",
      "WARN\tAdapter Content\tSRR2003569_1.fastq\n",
      "SRR2003569_2\n",
      "PASS\tBasic Statistics\tSRR2003569_2.fastq\n",
      "PASS\tPer base sequence quality\tSRR2003569_2.fastq\n",
      "PASS\tPer tile sequence quality\tSRR2003569_2.fastq\n",
      "PASS\tPer sequence quality scores\tSRR2003569_2.fastq\n",
      "WARN\tPer base sequence content\tSRR2003569_2.fastq\n",
      "PASS\tPer sequence GC content\tSRR2003569_2.fastq\n",
      "PASS\tPer base N content\tSRR2003569_2.fastq\n",
      "PASS\tSequence Length Distribution\tSRR2003569_2.fastq\n",
      "WARN\tSequence Duplication Levels\tSRR2003569_2.fastq\n",
      "PASS\tOverrepresented sequences\tSRR2003569_2.fastq\n",
      "WARN\tAdapter Content\tSRR2003569_2.fastq\n"
     ]
    }
   ],
   "source": [
    "##Unzip folders\n",
    "!unzip -qq \\*.zip\n",
    "print('SRR2003569_1')\n",
    "!cat SRR2003569_1_fastqc/summary.txt\n",
    "print('SRR2003569_2')\n",
    "!cat SRR2003569_2_fastqc/summary.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As seen in the summary, our reads passed most of the metrics for quality control. You can read the full report via the following links:\n",
    "\n",
    "<a href=\"analysis/FastQC/SRR2003569_1_fastqc.html\">SRR2003569_1</a>\n",
    "\n",
    "<a href=\"analysis/FastQC/SRR2003569_2_fastqc.html\">SRR2003569_2</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make directory for alignements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Move to analysis\n",
    "os.chdir(path + '/analysis/')\n",
    "##Create directory to place data\n",
    "!mkdir -p Alignments\n",
    "##Move to that directory\n",
    "os.chdir(path + '/analysis/Alignments')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (171, 889710, 298, 149)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (310, 856, 4026)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 11458)\n",
      "[M::mem_pestat] mean and std.dev: (2349.29, 2740.90)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 15174)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (136, 230, 283)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 577)\n",
      "[M::mem_pestat] mean and std.dev: (215.07, 93.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 724)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (211, 1563, 3810)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 11008)\n",
      "[M::mem_pestat] mean and std.dev: (2354.34, 2383.80)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14607)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (301, 867, 3348)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9442)\n",
      "[M::mem_pestat] mean and std.dev: (1878.43, 2126.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12489)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 296.380 CPU sec, 15.640 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (146, 891123, 320, 181)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (270, 760, 3789)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10827)\n",
      "[M::mem_pestat] mean and std.dev: (2161.57, 2636.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14346)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (131, 226, 280)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 578)\n",
      "[M::mem_pestat] mean and std.dev: (211.56, 93.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 727)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (215, 1348, 3393)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9749)\n",
      "[M::mem_pestat] mean and std.dev: (2130.56, 2296.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12927)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (281, 863, 2976)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8366)\n",
      "[M::mem_pestat] mean and std.dev: (1796.82, 2051.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11061)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 289.944 CPU sec, 15.208 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (149, 889289, 330, 169)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (339, 904, 3149)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8769)\n",
      "[M::mem_pestat] mean and std.dev: (1866.49, 2145.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11579)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (127, 221, 276)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 574)\n",
      "[M::mem_pestat] mean and std.dev: (207.72, 92.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 723)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (244, 1455, 3143)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8941)\n",
      "[M::mem_pestat] mean and std.dev: (2052.31, 2136.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11840)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (274, 727, 3101)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8755)\n",
      "[M::mem_pestat] mean and std.dev: (1749.84, 2073.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11582)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 294.296 CPU sec, 15.192 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (165, 891032, 308, 175)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (317, 862, 3587)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10127)\n",
      "[M::mem_pestat] mean and std.dev: (2373.96, 2750.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13397)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (134, 229, 282)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 578)\n",
      "[M::mem_pestat] mean and std.dev: (214.11, 93.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 726)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (182, 1361, 3138)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9050)\n",
      "[M::mem_pestat] mean and std.dev: (1981.51, 2170.61)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12006)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (272, 612, 2673)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7475)\n",
      "[M::mem_pestat] mean and std.dev: (1524.04, 1844.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9876)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 292.988 CPU sec, 14.974 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (191, 890251, 324, 181)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 680, 3452)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9882)\n",
      "[M::mem_pestat] mean and std.dev: (2184.20, 2731.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13108)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (130, 225, 279)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 577)\n",
      "[M::mem_pestat] mean and std.dev: (210.54, 93.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 726)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (298, 1583, 3424)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9676)\n",
      "[M::mem_pestat] mean and std.dev: (2257.66, 2307.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12802)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (356, 797, 2451)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 6641)\n",
      "[M::mem_pestat] mean and std.dev: (1327.09, 1440.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 8736)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 300.068 CPU sec, 15.172 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (160, 890609, 329, 163)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (277, 747, 3294)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9328)\n",
      "[M::mem_pestat] mean and std.dev: (1709.83, 2007.20)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12345)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (126, 220, 276)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 576)\n",
      "[M::mem_pestat] mean and std.dev: (207.07, 92.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 726)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 1574, 3533)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10131)\n",
      "[M::mem_pestat] mean and std.dev: (2322.01, 2407.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13430)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (322, 759, 2773)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7675)\n",
      "[M::mem_pestat] mean and std.dev: (1514.46, 1736.62)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10126)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 299.256 CPU sec, 15.621 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (214, 890506, 320, 159)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (292, 799, 2773)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7735)\n",
      "[M::mem_pestat] mean and std.dev: (1675.15, 1952.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10216)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (136, 230, 283)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 577)\n",
      "[M::mem_pestat] mean and std.dev: (214.89, 93.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 724)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (171, 1329, 2732)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7854)\n",
      "[M::mem_pestat] mean and std.dev: (1755.99, 1898.23)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10415)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (381, 932, 3734)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10440)\n",
      "[M::mem_pestat] mean and std.dev: (2399.70, 2758.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13793)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 295.416 CPU sec, 15.064 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (177, 890583, 291, 171)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (288, 796, 3595)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10209)\n",
      "[M::mem_pestat] mean and std.dev: (2122.62, 2529.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13516)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (131, 225, 279)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 575)\n",
      "[M::mem_pestat] mean and std.dev: (211.11, 92.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 723)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (192, 1541, 3247)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9357)\n",
      "[M::mem_pestat] mean and std.dev: (2129.18, 2286.90)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12412)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (270, 806, 3308)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9384)\n",
      "[M::mem_pestat] mean and std.dev: (1914.41, 2216.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12422)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 282.772 CPU sec, 14.535 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (179, 890210, 349, 175)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (288, 890, 2877)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8055)\n",
      "[M::mem_pestat] mean and std.dev: (1469.23, 1724.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10644)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (128, 221, 277)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 575)\n",
      "[M::mem_pestat] mean and std.dev: (208.44, 92.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 724)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (303, 1806, 3644)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10326)\n",
      "[M::mem_pestat] mean and std.dev: (2336.21, 2261.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13667)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (297, 733, 3727)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10587)\n",
      "[M::mem_pestat] mean and std.dev: (2153.62, 2596.63)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14017)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 289.452 CPU sec, 14.867 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (166, 890289, 311, 164)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (314, 757, 3373)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9491)\n",
      "[M::mem_pestat] mean and std.dev: (1983.41, 2356.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12550)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (136, 230, 283)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 577)\n",
      "[M::mem_pestat] mean and std.dev: (214.86, 93.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 724)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (224, 1543, 3704)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10664)\n",
      "[M::mem_pestat] mean and std.dev: (2285.67, 2315.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14144)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (311, 838, 2715)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7523)\n",
      "[M::mem_pestat] mean and std.dev: (1539.13, 1750.73)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9927)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 290.592 CPU sec, 14.673 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (165, 890071, 271, 180)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (276, 747, 3245)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9183)\n",
      "[M::mem_pestat] mean and std.dev: (1737.05, 2049.74)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12152)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (131, 225, 280)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 578)\n",
      "[M::mem_pestat] mean and std.dev: (211.26, 93.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 727)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (238, 1377, 3082)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8770)\n",
      "[M::mem_pestat] mean and std.dev: (2022.83, 2140.62)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11614)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (269, 765, 3305)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9377)\n",
      "[M::mem_pestat] mean and std.dev: (1876.74, 2295.73)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12413)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 296.596 CPU sec, 15.434 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (160, 890261, 341, 155)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (257, 892, 3212)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9122)\n",
      "[M::mem_pestat] mean and std.dev: (1956.46, 2254.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12077)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (129, 223, 278)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 576)\n",
      "[M::mem_pestat] mean and std.dev: (209.21, 92.61)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 725)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (192, 1334, 3023)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8685)\n",
      "[M::mem_pestat] mean and std.dev: (2000.84, 2228.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11516)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (330, 980, 2988)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8304)\n",
      "[M::mem_pestat] mean and std.dev: (1750.97, 1913.32)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10962)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 301.120 CPU sec, 15.374 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (171, 888897, 325, 184)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (409, 1215, 3715)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10327)\n",
      "[M::mem_pestat] mean and std.dev: (2302.18, 2294.21)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13633)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (133, 228, 282)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 580)\n",
      "[M::mem_pestat] mean and std.dev: (213.52, 93.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 729)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (164, 1378, 2932)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8468)\n",
      "[M::mem_pestat] mean and std.dev: (1913.73, 2107.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11236)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (329, 1033, 3421)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9605)\n",
      "[M::mem_pestat] mean and std.dev: (2026.42, 2253.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12697)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 292.784 CPU sec, 14.973 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (174, 889262, 296, 156)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (329, 1127, 3698)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10436)\n",
      "[M::mem_pestat] mean and std.dev: (2250.77, 2459.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13805)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (128, 223, 279)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 581)\n",
      "[M::mem_pestat] mean and std.dev: (209.66, 93.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 732)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (188, 1385, 3190)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9194)\n",
      "[M::mem_pestat] mean and std.dev: (2051.63, 2235.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12196)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (215, 639, 2789)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7937)\n",
      "[M::mem_pestat] mean and std.dev: (1418.68, 1801.90)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10511)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 295.712 CPU sec, 15.083 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (151, 890192, 300, 184)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (344, 1189, 3975)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 11237)\n",
      "[M::mem_pestat] mean and std.dev: (2515.18, 2796.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14868)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (127, 222, 277)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 577)\n",
      "[M::mem_pestat] mean and std.dev: (208.35, 92.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 727)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (194, 1418, 3354)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9674)\n",
      "[M::mem_pestat] mean and std.dev: (2144.23, 2310.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12834)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (366, 936, 3537)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9879)\n",
      "[M::mem_pestat] mean and std.dev: (2198.17, 2619.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13050)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mem_process_seqs] Processed 2222224 reads in 308.648 CPU sec, 15.747 real sec\n",
      "[M::process] read 2222224 sequences (200000160 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (167, 889988, 332, 174)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (331, 841, 2575)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7063)\n",
      "[M::mem_pestat] mean and std.dev: (1476.18, 1589.56)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9307)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (135, 229, 282)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 576)\n",
      "[M::mem_pestat] mean and std.dev: (214.48, 93.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 723)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (176, 1229, 2666)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7646)\n",
      "[M::mem_pestat] mean and std.dev: (1652.39, 1732.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10136)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 805, 2847)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7887)\n",
      "[M::mem_pestat] mean and std.dev: (1644.04, 1891.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 10407)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 294.968 CPU sec, 15.116 real sec\n",
      "[M::process] read 1812030 sequences (163082700 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (182, 889240, 317, 169)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (300, 783, 3002)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 8406)\n",
      "[M::mem_pestat] mean and std.dev: (1655.94, 1946.95)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11108)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (130, 225, 279)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 577)\n",
      "[M::mem_pestat] mean and std.dev: (210.92, 93.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 726)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (176, 1137, 3289)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9515)\n",
      "[M::mem_pestat] mean and std.dev: (2049.10, 2336.38)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 12628)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 809, 3735)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10727)\n",
      "[M::mem_pestat] mean and std.dev: (2247.62, 2749.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 14223)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2222224 reads in 293.380 CPU sec, 14.944 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (153, 725852, 258, 161)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (283, 682, 2695)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 7519)\n",
      "[M::mem_pestat] mean and std.dev: (1260.95, 1531.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 9931)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (128, 221, 276)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 572)\n",
      "[M::mem_pestat] mean and std.dev: (208.22, 92.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 720)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (225, 1436, 3536)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 10158)\n",
      "[M::mem_pestat] mean and std.dev: (2202.56, 2272.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 13469)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (342, 781, 3229)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 9003)\n",
      "[M::mem_pestat] mean and std.dev: (1827.63, 2168.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 11890)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 1812030 reads in 245.796 CPU sec, 12.897 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -t 20 -M /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/SRR2003569_1.fastq /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/SRR2003569_2.fastq\n",
      "[main] Real time: 276.060 sec; CPU: 5263.076 sec\n"
     ]
    }
   ],
   "source": [
    "!bwa mem -t {Ncpu} -M {ReferenceGenome} {FastqF1} {FastqF2} > {SampleName}.sam "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets transform to bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 20 in-memory blocks...\r\n"
     ]
    }
   ],
   "source": [
    "!samtools view -@ {Ncpu} -S -bh {SampleName}.sam | samtools sort -@ {Ncpu} - > {SampleName}.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools index {SampleName}.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter for mapping quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools view -@ {Ncpu} -q {minMaqQforBam} -bh {SampleName}.bam > {SampleName}.SF.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then add and replace groups via picard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2021-06-17 10:54:28\tAddOrReplaceReadGroups\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    AddOrReplaceReadGroups -I CB4856.SF.bam -O CB4856.RG.bam -RGID CB4856 -RGLB LB -RGPL illumina -RGPU PU -RGSM CB4856\n",
      "**********\n",
      "\n",
      "\n",
      "10:54:29.062 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/picard2.23.6/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Thu Jun 17 10:54:29 AST 2021] AddOrReplaceReadGroups INPUT=CB4856.SF.bam OUTPUT=CB4856.RG.bam RGID=CB4856 RGLB=LB RGPL=illumina RGPU=PU RGSM=CB4856    VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Thu Jun 17 10:54:29 AST 2021] Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.23.6\n",
      "INFO\t2021-06-17 10:54:29\tAddOrReplaceReadGroups\tCreated read-group ID=CB4856 PL=illumina LB=LB SM=CB4856\n",
      "\n",
      "INFO\t2021-06-17 10:54:39\tAddOrReplaceReadGroups\tProcessed     1,000,000 records.  Elapsed time: 00:00:10s.  Time for last 1,000,000:   10s.  Last read position: I:3,847,481\n",
      "INFO\t2021-06-17 10:54:48\tAddOrReplaceReadGroups\tProcessed     2,000,000 records.  Elapsed time: 00:00:19s.  Time for last 1,000,000:    9s.  Last read position: I:6,554,236\n",
      "INFO\t2021-06-17 10:54:58\tAddOrReplaceReadGroups\tProcessed     3,000,000 records.  Elapsed time: 00:00:29s.  Time for last 1,000,000:    9s.  Last read position: I:9,150,409\n",
      "INFO\t2021-06-17 10:55:07\tAddOrReplaceReadGroups\tProcessed     4,000,000 records.  Elapsed time: 00:00:38s.  Time for last 1,000,000:    9s.  Last read position: I:12,001,288\n",
      "INFO\t2021-06-17 10:55:17\tAddOrReplaceReadGroups\tProcessed     5,000,000 records.  Elapsed time: 00:00:48s.  Time for last 1,000,000:    9s.  Last read position: I:14,996,252\n",
      "INFO\t2021-06-17 10:55:26\tAddOrReplaceReadGroups\tProcessed     6,000,000 records.  Elapsed time: 00:00:57s.  Time for last 1,000,000:    9s.  Last read position: II:1,819,178\n",
      "INFO\t2021-06-17 10:55:36\tAddOrReplaceReadGroups\tProcessed     7,000,000 records.  Elapsed time: 00:01:07s.  Time for last 1,000,000:    9s.  Last read position: II:4,867,164\n",
      "INFO\t2021-06-17 10:55:47\tAddOrReplaceReadGroups\tProcessed     8,000,000 records.  Elapsed time: 00:01:18s.  Time for last 1,000,000:   10s.  Last read position: II:7,319,335\n",
      "INFO\t2021-06-17 10:55:56\tAddOrReplaceReadGroups\tProcessed     9,000,000 records.  Elapsed time: 00:01:27s.  Time for last 1,000,000:    9s.  Last read position: II:9,638,258\n",
      "INFO\t2021-06-17 10:56:06\tAddOrReplaceReadGroups\tProcessed    10,000,000 records.  Elapsed time: 00:01:37s.  Time for last 1,000,000:    9s.  Last read position: II:12,273,569\n",
      "INFO\t2021-06-17 10:56:16\tAddOrReplaceReadGroups\tProcessed    11,000,000 records.  Elapsed time: 00:01:47s.  Time for last 1,000,000:    9s.  Last read position: III:446,865\n",
      "INFO\t2021-06-17 10:56:25\tAddOrReplaceReadGroups\tProcessed    12,000,000 records.  Elapsed time: 00:01:56s.  Time for last 1,000,000:    9s.  Last read position: III:4,134,103\n",
      "INFO\t2021-06-17 10:56:35\tAddOrReplaceReadGroups\tProcessed    13,000,000 records.  Elapsed time: 00:02:06s.  Time for last 1,000,000:    9s.  Last read position: III:6,665,836\n",
      "INFO\t2021-06-17 10:56:44\tAddOrReplaceReadGroups\tProcessed    14,000,000 records.  Elapsed time: 00:02:15s.  Time for last 1,000,000:    9s.  Last read position: III:9,219,950\n",
      "INFO\t2021-06-17 10:56:54\tAddOrReplaceReadGroups\tProcessed    15,000,000 records.  Elapsed time: 00:02:25s.  Time for last 1,000,000:    9s.  Last read position: III:12,613,517\n",
      "INFO\t2021-06-17 10:57:04\tAddOrReplaceReadGroups\tProcessed    16,000,000 records.  Elapsed time: 00:02:35s.  Time for last 1,000,000:    9s.  Last read position: IV:2,474,211\n",
      "INFO\t2021-06-17 10:57:13\tAddOrReplaceReadGroups\tProcessed    17,000,000 records.  Elapsed time: 00:02:44s.  Time for last 1,000,000:    9s.  Last read position: IV:5,608,032\n",
      "INFO\t2021-06-17 10:57:23\tAddOrReplaceReadGroups\tProcessed    18,000,000 records.  Elapsed time: 00:02:54s.  Time for last 1,000,000:    9s.  Last read position: IV:8,297,527\n",
      "INFO\t2021-06-17 10:57:33\tAddOrReplaceReadGroups\tProcessed    19,000,000 records.  Elapsed time: 00:03:04s.  Time for last 1,000,000:   10s.  Last read position: IV:10,687,209\n",
      "INFO\t2021-06-17 10:57:43\tAddOrReplaceReadGroups\tProcessed    20,000,000 records.  Elapsed time: 00:03:13s.  Time for last 1,000,000:    9s.  Last read position: IV:13,257,819\n",
      "INFO\t2021-06-17 10:57:52\tAddOrReplaceReadGroups\tProcessed    21,000,000 records.  Elapsed time: 00:03:23s.  Time for last 1,000,000:    9s.  Last read position: IV:17,022,248\n",
      "INFO\t2021-06-17 10:58:02\tAddOrReplaceReadGroups\tProcessed    22,000,000 records.  Elapsed time: 00:03:33s.  Time for last 1,000,000:    9s.  Last read position: V:2,639,972\n",
      "INFO\t2021-06-17 10:58:12\tAddOrReplaceReadGroups\tProcessed    23,000,000 records.  Elapsed time: 00:03:43s.  Time for last 1,000,000:    9s.  Last read position: V:5,682,199\n",
      "INFO\t2021-06-17 10:58:22\tAddOrReplaceReadGroups\tProcessed    24,000,000 records.  Elapsed time: 00:03:53s.  Time for last 1,000,000:    9s.  Last read position: V:8,189,306\n",
      "INFO\t2021-06-17 10:58:32\tAddOrReplaceReadGroups\tProcessed    25,000,000 records.  Elapsed time: 00:04:02s.  Time for last 1,000,000:    9s.  Last read position: V:10,651,728\n",
      "INFO\t2021-06-17 10:58:41\tAddOrReplaceReadGroups\tProcessed    26,000,000 records.  Elapsed time: 00:04:12s.  Time for last 1,000,000:    9s.  Last read position: V:13,019,378\n",
      "INFO\t2021-06-17 10:58:50\tAddOrReplaceReadGroups\tProcessed    27,000,000 records.  Elapsed time: 00:04:21s.  Time for last 1,000,000:    9s.  Last read position: V:15,610,563\n",
      "INFO\t2021-06-17 10:59:01\tAddOrReplaceReadGroups\tProcessed    28,000,000 records.  Elapsed time: 00:04:31s.  Time for last 1,000,000:   10s.  Last read position: V:18,744,249\n",
      "INFO\t2021-06-17 10:59:11\tAddOrReplaceReadGroups\tProcessed    29,000,000 records.  Elapsed time: 00:04:41s.  Time for last 1,000,000:    9s.  Last read position: X:1,206,981\n",
      "INFO\t2021-06-17 10:59:20\tAddOrReplaceReadGroups\tProcessed    30,000,000 records.  Elapsed time: 00:04:51s.  Time for last 1,000,000:    9s.  Last read position: X:3,887,944\n",
      "INFO\t2021-06-17 10:59:30\tAddOrReplaceReadGroups\tProcessed    31,000,000 records.  Elapsed time: 00:05:01s.  Time for last 1,000,000:   10s.  Last read position: X:6,408,297\n",
      "INFO\t2021-06-17 10:59:40\tAddOrReplaceReadGroups\tProcessed    32,000,000 records.  Elapsed time: 00:05:11s.  Time for last 1,000,000:    9s.  Last read position: X:8,513,660\n",
      "INFO\t2021-06-17 10:59:49\tAddOrReplaceReadGroups\tProcessed    33,000,000 records.  Elapsed time: 00:05:20s.  Time for last 1,000,000:    9s.  Last read position: X:10,876,170\n",
      "INFO\t2021-06-17 10:59:59\tAddOrReplaceReadGroups\tProcessed    34,000,000 records.  Elapsed time: 00:05:30s.  Time for last 1,000,000:    9s.  Last read position: X:13,172,623\n",
      "INFO\t2021-06-17 11:00:09\tAddOrReplaceReadGroups\tProcessed    35,000,000 records.  Elapsed time: 00:05:40s.  Time for last 1,000,000:    9s.  Last read position: X:15,649,812\n",
      "[Thu Jun 17 11:00:17 AST 2021] picard.sam.AddOrReplaceReadGroups done. Elapsed time: 5.81 minutes.\n",
      "Runtime.totalMemory()=1681915904\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/picard2.23.6/picard.jar AddOrReplaceReadGroups I={SampleName}.SF.bam O={SampleName}.RG.bam RGID={SampleName} RGLB=LB RGPL=illumina RGPU=PU RGSM={SampleName}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we mark duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2021-06-17 11:01:10\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -I CB4856.RG.bam -O CB4856.Dup.bam -M CB4856.dedupMetrics -REMOVE_DUPLICATES true\n",
      "**********\n",
      "\n",
      "\n",
      "11:01:11.116 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/picard2.23.6/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Thu Jun 17 11:01:11 AST 2021] MarkDuplicates INPUT=[CB4856.RG.bam] OUTPUT=CB4856.Dup.bam METRICS_FILE=CB4856.dedupMetrics REMOVE_DUPLICATES=true    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Thu Jun 17 11:01:11 AST 2021] Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.23.6\n",
      "INFO\t2021-06-17 11:01:11\tMarkDuplicates\tStart of doWork freeMemory: 1997829648; totalMemory: 2020081664; maxMemory: 38177603584\n",
      "INFO\t2021-06-17 11:01:11\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2021-06-17 11:01:11\tMarkDuplicates\tWill retain up to 138324650 data points before spilling to disk.\n",
      "WARNING\t2021-06-17 11:01:12\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: SRR2003569.8879836. Cause: String 'SRR2003569.8879836' did not start with a parsable number.\n",
      "INFO\t2021-06-17 11:01:19\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:06s.  Time for last 1,000,000:    6s.  Last read position: I:3,847,481\n",
      "INFO\t2021-06-17 11:01:19\tMarkDuplicates\tTracking 8738 as yet unmatched pairs. 5332 records in RAM.\n",
      "INFO\t2021-06-17 11:01:27\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:14s.  Time for last 1,000,000:    8s.  Last read position: I:6,554,236\n",
      "INFO\t2021-06-17 11:01:27\tMarkDuplicates\tTracking 15413 as yet unmatched pairs. 11374 records in RAM.\n",
      "INFO\t2021-06-17 11:01:32\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:20s.  Time for last 1,000,000:    5s.  Last read position: I:9,150,409\n",
      "INFO\t2021-06-17 11:01:32\tMarkDuplicates\tTracking 16514 as yet unmatched pairs. 11930 records in RAM.\n",
      "INFO\t2021-06-17 11:01:39\tMarkDuplicates\tRead     4,000,000 records.  Elapsed time: 00:00:27s.  Time for last 1,000,000:    6s.  Last read position: I:12,001,288\n",
      "INFO\t2021-06-17 11:01:39\tMarkDuplicates\tTracking 22246 as yet unmatched pairs. 16587 records in RAM.\n",
      "INFO\t2021-06-17 11:01:43\tMarkDuplicates\tRead     5,000,000 records.  Elapsed time: 00:00:30s.  Time for last 1,000,000:    3s.  Last read position: I:14,996,252\n",
      "INFO\t2021-06-17 11:01:43\tMarkDuplicates\tTracking 27233 as yet unmatched pairs. 19200 records in RAM.\n",
      "INFO\t2021-06-17 11:01:50\tMarkDuplicates\tRead     6,000,000 records.  Elapsed time: 00:00:38s.  Time for last 1,000,000:    7s.  Last read position: II:1,819,178\n",
      "INFO\t2021-06-17 11:01:50\tMarkDuplicates\tTracking 32615 as yet unmatched pairs. 5318 records in RAM.\n",
      "INFO\t2021-06-17 11:01:56\tMarkDuplicates\tRead     7,000,000 records.  Elapsed time: 00:00:44s.  Time for last 1,000,000:    6s.  Last read position: II:4,867,164\n",
      "INFO\t2021-06-17 11:01:56\tMarkDuplicates\tTracking 39488 as yet unmatched pairs. 8945 records in RAM.\n",
      "INFO\t2021-06-17 11:02:00\tMarkDuplicates\tRead     8,000,000 records.  Elapsed time: 00:00:47s.  Time for last 1,000,000:    3s.  Last read position: II:7,319,335\n",
      "INFO\t2021-06-17 11:02:00\tMarkDuplicates\tTracking 41412 as yet unmatched pairs. 10378 records in RAM.\n",
      "INFO\t2021-06-17 11:02:18\tMarkDuplicates\tRead     9,000,000 records.  Elapsed time: 00:01:06s.  Time for last 1,000,000:   18s.  Last read position: II:9,638,258\n",
      "INFO\t2021-06-17 11:02:18\tMarkDuplicates\tTracking 43669 as yet unmatched pairs. 12280 records in RAM.\n",
      "INFO\t2021-06-17 11:02:23\tMarkDuplicates\tRead    10,000,000 records.  Elapsed time: 00:01:10s.  Time for last 1,000,000:    4s.  Last read position: II:12,273,569\n",
      "INFO\t2021-06-17 11:02:23\tMarkDuplicates\tTracking 46172 as yet unmatched pairs. 14123 records in RAM.\n",
      "INFO\t2021-06-17 11:02:28\tMarkDuplicates\tRead    11,000,000 records.  Elapsed time: 00:01:15s.  Time for last 1,000,000:    5s.  Last read position: III:446,865\n",
      "INFO\t2021-06-17 11:02:28\tMarkDuplicates\tTracking 52641 as yet unmatched pairs. 3802 records in RAM.\n",
      "INFO\t2021-06-17 11:02:31\tMarkDuplicates\tRead    12,000,000 records.  Elapsed time: 00:01:18s.  Time for last 1,000,000:    3s.  Last read position: III:4,134,103\n",
      "INFO\t2021-06-17 11:02:31\tMarkDuplicates\tTracking 60435 as yet unmatched pairs. 9055 records in RAM.\n",
      "INFO\t2021-06-17 11:02:39\tMarkDuplicates\tRead    13,000,000 records.  Elapsed time: 00:01:27s.  Time for last 1,000,000:    8s.  Last read position: III:6,665,836\n",
      "INFO\t2021-06-17 11:02:39\tMarkDuplicates\tTracking 61987 as yet unmatched pairs. 10036 records in RAM.\n",
      "INFO\t2021-06-17 11:02:42\tMarkDuplicates\tRead    14,000,000 records.  Elapsed time: 00:01:30s.  Time for last 1,000,000:    3s.  Last read position: III:9,219,950\n",
      "INFO\t2021-06-17 11:02:42\tMarkDuplicates\tTracking 63605 as yet unmatched pairs. 11244 records in RAM.\n",
      "INFO\t2021-06-17 11:02:45\tMarkDuplicates\tRead    15,000,000 records.  Elapsed time: 00:01:33s.  Time for last 1,000,000:    3s.  Last read position: III:12,613,517\n",
      "INFO\t2021-06-17 11:02:45\tMarkDuplicates\tTracking 67702 as yet unmatched pairs. 13851 records in RAM.\n",
      "INFO\t2021-06-17 11:02:51\tMarkDuplicates\tRead    16,000,000 records.  Elapsed time: 00:01:38s.  Time for last 1,000,000:    5s.  Last read position: IV:2,474,211\n",
      "INFO\t2021-06-17 11:02:51\tMarkDuplicates\tTracking 71436 as yet unmatched pairs. 7118 records in RAM.\n",
      "INFO\t2021-06-17 11:02:54\tMarkDuplicates\tRead    17,000,000 records.  Elapsed time: 00:01:41s.  Time for last 1,000,000:    3s.  Last read position: IV:5,608,032\n",
      "INFO\t2021-06-17 11:02:54\tMarkDuplicates\tTracking 79833 as yet unmatched pairs. 14313 records in RAM.\n",
      "INFO\t2021-06-17 11:02:57\tMarkDuplicates\tRead    18,000,000 records.  Elapsed time: 00:01:44s.  Time for last 1,000,000:    3s.  Last read position: IV:8,297,527\n",
      "INFO\t2021-06-17 11:02:57\tMarkDuplicates\tTracking 83064 as yet unmatched pairs. 16816 records in RAM.\n",
      "INFO\t2021-06-17 11:03:07\tMarkDuplicates\tRead    19,000,000 records.  Elapsed time: 00:01:54s.  Time for last 1,000,000:   10s.  Last read position: IV:10,687,209\n",
      "INFO\t2021-06-17 11:03:07\tMarkDuplicates\tTracking 90029 as yet unmatched pairs. 23192 records in RAM.\n",
      "INFO\t2021-06-17 11:03:10\tMarkDuplicates\tRead    20,000,000 records.  Elapsed time: 00:01:58s.  Time for last 1,000,000:    3s.  Last read position: IV:13,257,819\n",
      "INFO\t2021-06-17 11:03:10\tMarkDuplicates\tTracking 98465 as yet unmatched pairs. 31165 records in RAM.\n",
      "INFO\t2021-06-17 11:03:13\tMarkDuplicates\tRead    21,000,000 records.  Elapsed time: 00:02:01s.  Time for last 1,000,000:    3s.  Last read position: IV:17,022,248\n",
      "INFO\t2021-06-17 11:03:13\tMarkDuplicates\tTracking 106339 as yet unmatched pairs. 37003 records in RAM.\n",
      "INFO\t2021-06-17 11:03:45\tMarkDuplicates\tRead    22,000,000 records.  Elapsed time: 00:02:33s.  Time for last 1,000,000:   31s.  Last read position: V:2,639,972\n",
      "INFO\t2021-06-17 11:03:45\tMarkDuplicates\tTracking 122847 as yet unmatched pairs. 24282 records in RAM.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2021-06-17 11:03:49\tMarkDuplicates\tRead    23,000,000 records.  Elapsed time: 00:02:37s.  Time for last 1,000,000:    4s.  Last read position: V:5,682,199\n",
      "INFO\t2021-06-17 11:03:49\tMarkDuplicates\tTracking 126877 as yet unmatched pairs. 26696 records in RAM.\n",
      "INFO\t2021-06-17 11:03:53\tMarkDuplicates\tRead    24,000,000 records.  Elapsed time: 00:02:41s.  Time for last 1,000,000:    3s.  Last read position: V:8,189,306\n",
      "INFO\t2021-06-17 11:03:53\tMarkDuplicates\tTracking 131254 as yet unmatched pairs. 30493 records in RAM.\n",
      "INFO\t2021-06-17 11:03:57\tMarkDuplicates\tRead    25,000,000 records.  Elapsed time: 00:02:45s.  Time for last 1,000,000:    3s.  Last read position: V:10,651,728\n",
      "INFO\t2021-06-17 11:03:57\tMarkDuplicates\tTracking 134890 as yet unmatched pairs. 33772 records in RAM.\n",
      "INFO\t2021-06-17 11:04:04\tMarkDuplicates\tRead    26,000,000 records.  Elapsed time: 00:02:52s.  Time for last 1,000,000:    7s.  Last read position: V:13,019,378\n",
      "INFO\t2021-06-17 11:04:04\tMarkDuplicates\tTracking 135782 as yet unmatched pairs. 34419 records in RAM.\n",
      "INFO\t2021-06-17 11:04:09\tMarkDuplicates\tRead    27,000,000 records.  Elapsed time: 00:02:56s.  Time for last 1,000,000:    4s.  Last read position: V:15,610,563\n",
      "INFO\t2021-06-17 11:04:09\tMarkDuplicates\tTracking 137749 as yet unmatched pairs. 35796 records in RAM.\n",
      "INFO\t2021-06-17 11:04:12\tMarkDuplicates\tRead    28,000,000 records.  Elapsed time: 00:02:59s.  Time for last 1,000,000:    3s.  Last read position: V:18,744,249\n",
      "INFO\t2021-06-17 11:04:12\tMarkDuplicates\tTracking 141678 as yet unmatched pairs. 37689 records in RAM.\n",
      "INFO\t2021-06-17 11:04:20\tMarkDuplicates\tRead    29,000,000 records.  Elapsed time: 00:03:07s.  Time for last 1,000,000:    7s.  Last read position: X:1,206,981\n",
      "INFO\t2021-06-17 11:04:20\tMarkDuplicates\tTracking 146910 as yet unmatched pairs. 9697 records in RAM.\n",
      "INFO\t2021-06-17 11:04:23\tMarkDuplicates\tRead    30,000,000 records.  Elapsed time: 00:03:10s.  Time for last 1,000,000:    3s.  Last read position: X:3,887,944\n",
      "INFO\t2021-06-17 11:04:23\tMarkDuplicates\tTracking 151108 as yet unmatched pairs. 13254 records in RAM.\n",
      "INFO\t2021-06-17 11:04:26\tMarkDuplicates\tRead    31,000,000 records.  Elapsed time: 00:03:13s.  Time for last 1,000,000:    3s.  Last read position: X:6,408,297\n",
      "INFO\t2021-06-17 11:04:26\tMarkDuplicates\tTracking 152503 as yet unmatched pairs. 14493 records in RAM.\n",
      "INFO\t2021-06-17 11:04:29\tMarkDuplicates\tRead    32,000,000 records.  Elapsed time: 00:03:16s.  Time for last 1,000,000:    3s.  Last read position: X:8,513,660\n",
      "INFO\t2021-06-17 11:04:29\tMarkDuplicates\tTracking 166908 as yet unmatched pairs. 28611 records in RAM.\n",
      "INFO\t2021-06-17 11:04:35\tMarkDuplicates\tRead    33,000,000 records.  Elapsed time: 00:03:22s.  Time for last 1,000,000:    6s.  Last read position: X:10,876,170\n",
      "INFO\t2021-06-17 11:04:35\tMarkDuplicates\tTracking 169271 as yet unmatched pairs. 30733 records in RAM.\n",
      "INFO\t2021-06-17 11:04:38\tMarkDuplicates\tRead    34,000,000 records.  Elapsed time: 00:03:26s.  Time for last 1,000,000:    3s.  Last read position: X:13,172,623\n",
      "INFO\t2021-06-17 11:04:38\tMarkDuplicates\tTracking 173092 as yet unmatched pairs. 34302 records in RAM.\n",
      "INFO\t2021-06-17 11:04:41\tMarkDuplicates\tRead    35,000,000 records.  Elapsed time: 00:03:29s.  Time for last 1,000,000:    3s.  Last read position: X:15,649,812\n",
      "INFO\t2021-06-17 11:04:41\tMarkDuplicates\tTracking 177312 as yet unmatched pairs. 37967 records in RAM.\n",
      "INFO\t2021-06-17 11:05:29\tMarkDuplicates\tRead 35894952 records. 182029 pairs never matched.\n",
      "INFO\t2021-06-17 11:06:10\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 15231525512; totalMemory: 22002794496; maxMemory: 38177603584\n",
      "INFO\t2021-06-17 11:06:10\tMarkDuplicates\tWill retain up to 1193050112 duplicate indices before spilling to disk.\n",
      "INFO\t2021-06-17 11:06:14\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2021-06-17 11:07:03\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2021-06-17 11:07:05\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2021-06-17 11:07:06\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 26462233440; totalMemory: 36247175168; maxMemory: 38177603584\n",
      "INFO\t2021-06-17 11:07:06\tMarkDuplicates\tMarking 1075668 records as duplicates.\n",
      "INFO\t2021-06-17 11:07:06\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2021-06-17 11:07:06\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2021-06-17 11:08:45\tMarkDuplicates\tWritten    10,000,000 records.  Elapsed time: 00:01:39s.  Time for last 10,000,000:   99s.  Last read position: II:13,606,866\n",
      "INFO\t2021-06-17 11:10:25\tMarkDuplicates\tWritten    20,000,000 records.  Elapsed time: 00:03:19s.  Time for last 10,000,000:   99s.  Last read position: IV:15,444,781\n",
      "INFO\t2021-06-17 11:12:05\tMarkDuplicates\tWritten    30,000,000 records.  Elapsed time: 00:04:59s.  Time for last 10,000,000:   99s.  Last read position: X:6,105,480\n",
      "INFO\t2021-06-17 11:12:52\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2021-06-17 11:12:52\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2021-06-17 11:12:52\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2021-06-17 11:12:52\tMarkDuplicates\tBefore output close freeMemory: 36678656424; totalMemory: 36942905344; maxMemory: 38177603584\n",
      "INFO\t2021-06-17 11:12:52\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2021-06-17 11:12:52\tMarkDuplicates\tAfter output close freeMemory: 36096172456; totalMemory: 36360421376; maxMemory: 38177603584\n",
      "[Thu Jun 17 11:12:52 AST 2021] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 11.70 minutes.\n",
      "Runtime.totalMemory()=36360421376\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/picard2.23.6/picard.jar MarkDuplicates I={SampleName}.RG.bam O={SampleName}.Dup.bam M={SampleName}.dedupMetrics REMOVE_DUPLICATES=true"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Index bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools index {SampleName}.Dup.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can realign\n",
    "\n",
    "#### GATK realigments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GATK needs dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2021-06-17 11:30:00\tCreateSequenceDictionary\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    CreateSequenceDictionary -R /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa -O /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.dict\n",
      "**********\n",
      "\n",
      "\n",
      "11:30:01.024 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/picard2.23.6/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Thu Jun 17 11:30:01 AST 2021] CreateSequenceDictionary OUTPUT=/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.dict REFERENCE=/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Thu Jun 17 11:30:01 AST 2021] Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.23.6\n",
      "[Thu Jun 17 11:30:02 AST 2021] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.02 minutes.\n",
      "Runtime.totalMemory()=2020081664\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/picard2.23.6/picard.jar CreateSequenceDictionary R={ReferenceGenome} O={ReferenceGenome[:-3]}.dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  13:18:28,146 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  13:18:28,149 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.8-1-0-gf15c1c3ef, Compiled 2018/02/19 05:43:50 \n",
      "INFO  13:18:28,149 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  13:18:28,149 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  13:18:28,149 HelpFormatter - [Thu Jun 17 13:18:28 AST 2021] Executing on Linux 4.4.0-210-generic amd64 \n",
      "INFO  13:18:28,149 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10 \n",
      "INFO  13:18:28,153 HelpFormatter - Program Args: -T RealignerTargetCreator -nt 20 -R /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa -I CB4856.Dup.bam -o CB4856.Indel.intervals \n",
      "INFO  13:18:28,158 HelpFormatter - Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10. \n",
      "INFO  13:18:28,159 HelpFormatter - Date/Time: 2021/06/17 13:18:28 \n",
      "INFO  13:18:28,159 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  13:18:28,159 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  13:18:28,171 NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar!/com/intel/gkl/native/libgkl_compression.so \n",
      "INFO  13:18:28,201 GenomeAnalysisEngine - Deflater: IntelDeflater \n",
      "INFO  13:18:28,201 GenomeAnalysisEngine - Inflater: IntelInflater \n",
      "INFO  13:18:28,202 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  13:18:28,552 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  13:18:28,559 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,581 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.02 \n",
      "INFO  13:18:28,605 MicroScheduler - Running the GATK in parallel mode with 20 total threads, 1 CPU thread(s) for each of 20 data thread(s), of 32 processors available on this machine \n",
      "INFO  13:18:28,681 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  13:18:28,728 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  13:18:28,729 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  13:18:28,729 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  13:18:28,729 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  13:18:28,747 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,756 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,759 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,765 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,766 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,775 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,776 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,783 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,784 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,794 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,795 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,804 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,805 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,811 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,812 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,819 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,820 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,827 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,828 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,833 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,834 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,843 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,844 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,848 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  13:18:28,849 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,858 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,858 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,868 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,868 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,876 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,879 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,887 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,887 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,896 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,897 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,904 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:28,904 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:18:28,912 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  13:18:58,736 ProgressMeter -     MtDNA:13701   1.00272607E8    30.0 s       0.0 s      100.0%    30.0 s       0.0 s \n",
      "INFO  13:18:58,789 ProgressMeter -            done   1.00286401E8    30.0 s       0.0 s      100.0%    30.0 s       0.0 s \n",
      "INFO  13:18:58,789 ProgressMeter - Total runtime 30.06 secs, 0.50 min, 0.01 hours \n",
      "INFO  13:18:58,790 MicroScheduler - 78372 reads were filtered out during the traversal out of approximately 34995004 total reads (0.22%) \n",
      "INFO  13:18:58,791 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  13:18:58,791 MicroScheduler -   -> 59444 reads (0.17% of total) failing BadMateFilter \n",
      "INFO  13:18:58,792 MicroScheduler -   -> 0 reads (0.00% of total) failing DuplicateReadFilter \n",
      "INFO  13:18:58,792 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  13:18:58,792 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  13:18:58,792 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  13:18:58,793 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityZeroFilter \n",
      "INFO  13:18:58,793 MicroScheduler -   -> 18928 reads (0.05% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  13:18:58,793 MicroScheduler -   -> 0 reads (0.00% of total) failing Platform454Filter \n",
      "INFO  13:18:58,794 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were no warn messages.\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/gatk-3.8.1.0/GenomeAnalysisTK.jar -T RealignerTargetCreator -nt {Ncpu} -R {ReferenceGenome} -I {SampleName}.Dup.bam -o {SampleName}.Indel.intervals"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Realigment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  13:19:13,970 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  13:19:13,972 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.8-1-0-gf15c1c3ef, Compiled 2018/02/19 05:43:50 \n",
      "INFO  13:19:13,973 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  13:19:13,973 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  13:19:13,973 HelpFormatter - [Thu Jun 17 13:19:13 AST 2021] Executing on Linux 4.4.0-210-generic amd64 \n",
      "INFO  13:19:13,973 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10 \n",
      "INFO  13:19:13,978 HelpFormatter - Program Args: -T IndelRealigner -R /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa -I CB4856.Dup.bam -targetIntervals CB4856.Indel.intervals -o CB4856.realigned.bam \n",
      "INFO  13:19:13,986 HelpFormatter - Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10. \n",
      "INFO  13:19:13,986 HelpFormatter - Date/Time: 2021/06/17 13:19:13 \n",
      "INFO  13:19:13,986 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  13:19:13,986 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  13:19:14,271 NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar!/com/intel/gkl/native/libgkl_compression.so \n",
      "INFO  13:19:14,303 GenomeAnalysisEngine - Deflater: IntelDeflater \n",
      "INFO  13:19:14,303 GenomeAnalysisEngine - Inflater: IntelInflater \n",
      "INFO  13:19:14,304 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  13:19:14,440 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  13:19:14,447 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  13:19:14,477 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.03 \n",
      "INFO  13:19:14,593 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  13:19:14,598 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  13:19:14,598 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  13:19:14,599 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  13:19:14,599 ProgressMeter -        Location |     reads | elapsed |     reads | completed | runtime |   runtime \n",
      "INFO  13:19:14,728 ReadShardBalancer$1 - Loading BAM index data \n",
      "INFO  13:19:14,730 ReadShardBalancer$1 - Done loading BAM index data \n",
      "INFO  13:19:44,606 ProgressMeter -       I:5730043   1600023.0    30.0 s      18.0 s        5.7%     8.8 m       8.3 m \n",
      "INFO  13:20:14,610 ProgressMeter -      I:10946559   3500095.0    60.0 s      17.0 s       10.9%     9.2 m       8.2 m \n",
      "INFO  13:20:44,614 ProgressMeter -      I:15066408   5000126.0    90.0 s      18.0 s       15.0%    10.0 m       8.5 m \n",
      "INFO  13:21:14,640 ProgressMeter -      II:4445771   6422764.0   120.0 s      18.0 s       19.5%    10.3 m       8.3 m \n",
      "INFO  13:21:44,642 ProgressMeter -      II:9082352   8322799.0     2.5 m      18.0 s       24.1%    10.4 m       7.9 m \n",
      "INFO  13:22:14,646 ProgressMeter -     II:14587420   1.0222834E7     3.0 m      17.0 s       29.6%    10.1 m       7.1 m \n",
      "INFO  13:22:44,650 ProgressMeter -     III:5113279   1.1883053E7     3.5 m      17.0 s       35.4%     9.9 m       6.4 m \n",
      "INFO  13:23:14,654 ProgressMeter -    III:10534247   1.3983088E7     4.0 m      17.0 s       40.8%     9.8 m       5.8 m \n",
      "INFO  13:23:44,658 ProgressMeter -      IV:3583713   1.5746813E7     4.5 m      17.0 s       47.6%     9.5 m       5.0 m \n",
      "INFO  13:24:14,662 ProgressMeter -      IV:8470530   1.744684E7     5.0 m      17.0 s       52.5%     9.5 m       4.5 m \n",
      "INFO  13:24:44,665 ProgressMeter -     IV:11904534   1.8846866E7     5.5 m      17.0 s       55.9%     9.8 m       4.3 m \n",
      "INFO  13:25:14,667 ProgressMeter -        V:265481   2.0507808E7     6.0 m      17.0 s       61.7%     9.7 m       3.7 m \n",
      "INFO  13:25:44,671 ProgressMeter -       V:1649966   2.1007817E7     6.5 m      18.0 s       63.1%    10.3 m       3.8 m \n",
      "INFO  13:26:14,673 ProgressMeter -       V:7542456   2.300786E7     7.0 m      18.0 s       69.0%    10.1 m       3.1 m \n",
      "INFO  13:26:44,675 ProgressMeter -      V:12518350   2.5007888E7     7.5 m      17.0 s       73.9%    10.1 m       2.6 m \n",
      "INFO  13:27:16,501 ProgressMeter -      V:17915251   2.6907916E7     8.0 m      17.0 s       79.3%    10.1 m       2.1 m \n",
      "INFO  13:27:46,504 ProgressMeter -       X:2444815   2.8579741E7     8.5 m      17.0 s       84.8%    10.0 m      91.0 s \n",
      "INFO  13:28:16,508 ProgressMeter -       X:7170292   3.0379773E7     9.0 m      17.0 s       89.5%    10.1 m      63.0 s \n",
      "INFO  13:28:46,512 ProgressMeter -      X:11891184   3.2379816E7     9.5 m      17.0 s       94.2%    10.1 m      35.0 s \n",
      "INFO  13:29:16,516 ProgressMeter -      X:15811185   3.3979867E7    10.0 m      17.0 s       98.1%    10.2 m      11.0 s \n",
      "INFO  13:29:31,510 ProgressMeter -            done   3.4819284E7    10.3 m      17.0 s      100.0%    10.3 m       0.0 s \n",
      "INFO  13:29:31,510 ProgressMeter - Total runtime 615.09 secs, 10.25 min, 0.17 hours \n",
      "INFO  13:29:31,513 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 34819284 total reads (0.00%) \n",
      "INFO  13:29:31,514 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  13:29:31,514 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were no warn messages.\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/gatk-3.8.1.0/GenomeAnalysisTK.jar -T IndelRealigner -R {ReferenceGenome} -I {SampleName}.Dup.bam -targetIntervals {SampleName}.Indel.intervals -o {SampleName}.realigned.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check files present and remove temporal files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB4856.bam.bai\t     CB4856.Indel.intervals\tCB4856.realigned.bam\r\n",
      "CB4856.dedupMetrics  CB4856.IndelIntervals.txt\r\n",
      "CB4856.Dup.bam.bai   CB4856.realigned.bai\r\n"
     ]
    }
   ],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm {SampleName}.sam\n",
    "!rm {SampleName}.bam\n",
    "!rm {SampleName}.SF.bam\n",
    "!rm {SampleName}.RG.bam\n",
    "!rm {SampleName}.Dup.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now is time for variant calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Variant calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Move to analysis\n",
    "os.chdir(path + '/analysis/')\n",
    "##Create directory to place data\n",
    "!mkdir -p VariantCalling\n",
    "##Move to that directory\n",
    "os.chdir(path + '/analysis/VariantCalling')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use GATK unified genotyper thouhg most recent versions uses Haplotype callers (it should be better) due to actual base recalibration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  17:02:12,156 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  17:02:12,158 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.8-1-0-gf15c1c3ef, Compiled 2018/02/19 05:43:50 \n",
      "INFO  17:02:12,158 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  17:02:12,158 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  17:02:12,158 HelpFormatter - [Thu Jun 17 17:02:12 AST 2021] Executing on Linux 4.4.0-210-generic amd64 \n",
      "INFO  17:02:12,158 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10 \n",
      "INFO  17:02:12,161 HelpFormatter - Program Args: -T UnifiedGenotyper -R /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa -nt 20 -l INFO -glm BOTH -I /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/analysis/Alignments/CB4856.realigned.bam -o CB4856.UG.vcf -mbq 10 -stand_call_conf 10 \n",
      "INFO  17:02:12,167 HelpFormatter - Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10. \n",
      "INFO  17:02:12,167 HelpFormatter - Date/Time: 2021/06/17 17:02:12 \n",
      "INFO  17:02:12,167 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  17:02:12,167 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  17:02:12,184 NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar!/com/intel/gkl/native/libgkl_compression.so \n",
      "INFO  17:02:12,217 GenomeAnalysisEngine - Deflater: IntelDeflater \n",
      "INFO  17:02:12,218 GenomeAnalysisEngine - Inflater: IntelInflater \n",
      "INFO  17:02:12,218 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  17:02:12,552 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 250 \n",
      "INFO  17:02:12,558 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,573 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,593 MicroScheduler - Running the GATK in parallel mode with 20 total threads, 1 CPU thread(s) for each of 20 data thread(s), of 32 processors available on this machine \n",
      "INFO  17:02:12,651 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  17:02:12,696 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  17:02:12,697 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  17:02:12,697 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  17:02:12,697 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  17:02:12,747 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "WARN  17:02:12,747 InbreedingCoeff - Annotation will not be calculated. InbreedingCoeff requires at least 10 unrelated samples. \n",
      "INFO  17:02:12,748 StrandBiasTest - SAM/BAM data was found. Attempting to use read data to calculate strand bias annotations values. \n",
      "INFO  17:02:12,819 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,827 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,830 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,836 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,841 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,848 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,851 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,859 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,864 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,868 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  17:02:12,877 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,880 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  17:02:12,891 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,894 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  17:02:12,898 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,904 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,908 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,911 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  17:02:12,915 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,920 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  17:02:12,922 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,927 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,935 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,955 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.02 \n",
      "INFO  17:02:12,961 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,964 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  17:02:12,970 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,979 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:12,988 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:12,992 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  17:02:12,998 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:13,011 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:13,017 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:13,024 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:13,034 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:13,044 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:13,045 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  17:02:13,055 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  17:02:42,704 ProgressMeter -     II:10960997   2.5754802E7    30.0 s       1.0 s       26.0%   115.0 s      85.0 s \n",
      "INFO  17:03:12,706 ProgressMeter -     IV:15515649   5.9340008E7    60.0 s       1.0 s       59.5%   100.0 s      40.0 s \n",
      "INFO  17:03:42,709 ProgressMeter -       X:7750533   8.9992001E7    90.0 s       1.0 s       90.0%    99.0 s       9.0 s \n",
      "INFO  17:03:53,079 ProgressMeter -            done   1.00286401E8   100.0 s       1.0 s      100.0%   100.0 s       0.0 s \n",
      "INFO  17:03:53,079 ProgressMeter - Total runtime 100.38 secs, 1.67 min, 0.03 hours \n",
      "INFO  17:03:53,080 MicroScheduler - 78372 reads were filtered out during the traversal out of approximately 34995135 total reads (0.22%) \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 59444 reads (0.17% of total) failing BadMateFilter \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 0 reads (0.00% of total) failing DuplicateReadFilter \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 18928 reads (0.05% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  17:03:53,080 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were 1 WARN messages, the first 1 are repeated below.\n",
      "WARN  17:02:12,747 InbreedingCoeff - Annotation will not be calculated. InbreedingCoeff requires at least 10 unrelated samples. \n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/gatk-3.8.1.0/GenomeAnalysisTK.jar -T UnifiedGenotyper -R {ReferenceGenome} -nt {Ncpu} -l INFO -glm BOTH -I {path}/analysis/Alignments/{SampleName}.realigned.bam -o {SampleName}.UG.vcf -mbq {minBaseQforVcf} -stand_call_conf {minVarCall}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now lets try with HaplotypeCaller"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Emit confidence for vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  09:34:38,299 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  09:34:38,300 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.8-1-0-gf15c1c3ef, Compiled 2018/02/19 05:43:50 \n",
      "INFO  09:34:38,301 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  09:34:38,301 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  09:34:38,301 HelpFormatter - [Sun Jun 20 09:34:38 AST 2021] Executing on Linux 4.4.0-210-generic amd64 \n",
      "INFO  09:34:38,301 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10 \n",
      "INFO  09:34:38,304 HelpFormatter - Program Args: -T HaplotypeCaller --emitRefConfidence GVCF -R /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa -l INFO -I /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/analysis/Alignments/CB4856.realigned.bam -o CB4856.g.vcf -mbq 10 -mmq 10 -stand_call_conf 10 \n",
      "INFO  09:34:38,309 HelpFormatter - Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10. \n",
      "INFO  09:34:38,310 HelpFormatter - Date/Time: 2021/06/20 09:34:38 \n",
      "INFO  09:34:38,310 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  09:34:38,310 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  09:34:38,325 NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar!/com/intel/gkl/native/libgkl_compression.so \n",
      "INFO  09:34:38,352 GenomeAnalysisEngine - Deflater: IntelDeflater \n",
      "INFO  09:34:38,352 GenomeAnalysisEngine - Inflater: IntelInflater \n",
      "INFO  09:34:38,352 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  09:34:38,653 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 500 \n",
      "INFO  09:34:38,658 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  09:34:38,673 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.01 \n",
      "INFO  09:34:38,681 HCMappingQualityFilter - Filtering out reads with MAPQ < 10 \n",
      "INFO  09:34:38,756 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  09:34:38,801 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  09:34:38,801 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  09:34:38,801 ProgressMeter -                 |      processed |    time |         per 1M |           |   total | remaining \n",
      "INFO  09:34:38,802 ProgressMeter -        Location | active regions | elapsed | active regions | completed | runtime |   runtime \n",
      "INFO  09:34:38,802 HaplotypeCaller - Standard Emitting and Calling confidence set to 0.0 for reference-model confidence output \n",
      "INFO  09:34:38,803 HaplotypeCaller - All sites annotated with PLs forced to true for reference-model confidence output \n",
      "WARN  09:34:38,857 InbreedingCoeff - Annotation will not be calculated. InbreedingCoeff requires at least 10 unrelated samples. \n",
      "INFO  09:34:38,924 HaplotypeCaller - Using global mismapping rate of 45 => -4.5 in log10 likelihood units \n",
      "INFO  09:34:40,009 NativeLibraryLoader - Loading libgkl_utils.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar!/com/intel/gkl/native/libgkl_utils.so \n",
      "INFO  09:34:40,012 NativeLibraryLoader - Loading libgkl_pairhmm_omp.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar!/com/intel/gkl/native/libgkl_pairhmm_omp.so \n",
      "INFO  09:34:40,054 VectorLoglessPairHMM - Using OpenMP multi-threaded AVX-accelerated native PairHMM implementation \n",
      "INFO  09:34:40,054 IntelPairHmm - Using CPU-supported AVX-512 instructions \n",
      "WARN  09:34:40,054 IntelPairHmm - Flush-to-zero (FTZ) is enabled when running PairHMM \n",
      "INFO  09:34:40,054 IntelPairHmm - Available threads: 32 \n",
      "INFO  09:34:40,054 IntelPairHmm - Requested threads: 1 \n",
      "WARN  09:34:40,174 HaplotypeScore - Annotation will not be calculated, must be called from UnifiedGenotyper, not HaplotypeCaller \n",
      "WARN  09:34:40,429 AnnotationUtils - Annotation will not be calculated, genotype is not called \n",
      "WARN  09:34:40,429 AnnotationUtils - Annotation will not be calculated, genotype is not called \n",
      "INFO  09:35:08,808 ProgressMeter -        I:362829              0.0    30.0 s           49.6 w        0.4%     2.3 h       2.3 h \n",
      "INFO  09:35:38,810 ProgressMeter -        I:806638              0.0    60.0 s           99.2 w        0.8%     2.1 h       2.1 h \n",
      "INFO  09:36:08,813 ProgressMeter -       I:1162179              0.0    90.0 s          148.8 w        1.2%     2.2 h       2.1 h \n",
      "INFO  09:36:38,816 ProgressMeter -       I:1499262              0.0   120.0 s          198.4 w        1.5%     2.2 h       2.2 h \n",
      "INFO  09:37:08,818 ProgressMeter -       I:1852421              0.0     2.5 m          248.0 w        1.8%     2.3 h       2.2 h \n",
      "INFO  09:37:38,821 ProgressMeter -       I:2214808              0.0     3.0 m          297.7 w        2.2%     2.3 h       2.2 h \n",
      "INFO  09:38:08,824 ProgressMeter -       I:2658534              0.0     3.5 m          347.3 w        2.7%     2.2 h       2.1 h \n",
      "INFO  09:38:38,826 ProgressMeter -       I:3013743              0.0     4.0 m          396.9 w        3.0%     2.2 h       2.2 h \n",
      "INFO  09:39:08,828 ProgressMeter -       I:3556557              0.0     4.5 m          446.5 w        3.5%     2.1 h       2.0 h \n",
      "INFO  09:39:38,830 ProgressMeter -       I:4248641              0.0     5.0 m          496.1 w        4.2%   118.0 m     113.0 m \n",
      "INFO  09:40:08,832 ProgressMeter -       I:5025948              0.0     5.5 m          545.7 w        5.0%   109.7 m     104.2 m \n",
      "INFO  09:40:38,834 ProgressMeter -       I:5718035              0.0     6.0 m          595.3 w        5.7%   105.2 m      99.2 m \n",
      "INFO  09:41:10,619 ProgressMeter -       I:6451487              0.0     6.5 m          644.9 w        6.4%   101.0 m      94.5 m \n",
      "INFO  09:41:40,621 ProgressMeter -       I:7275475              0.0     7.0 m          694.5 w        7.3%    96.5 m      89.5 m \n",
      "INFO  09:42:10,622 ProgressMeter -       I:7957390              0.0     7.5 m          744.1 w        7.9%    94.5 m      87.0 m \n",
      "INFO  09:42:40,624 ProgressMeter -       I:8576446              0.0     8.0 m          793.7 w        8.6%    93.5 m      85.5 m \n",
      "INFO  09:43:10,627 ProgressMeter -       I:9287344              0.0     8.5 m          843.3 w        9.3%    91.8 m      83.3 m \n",
      "INFO  09:43:40,629 ProgressMeter -       I:9928643              0.0     9.0 m          892.9 w        9.9%    90.9 m      81.9 m \n",
      "INFO  09:44:10,631 ProgressMeter -      I:10571502              0.0     9.5 m          942.5 w       10.5%    90.1 m      80.6 m \n",
      "INFO  09:44:40,634 ProgressMeter -      I:11049921              0.0    10.0 m          992.1 w       11.0%    90.8 m      80.8 m \n",
      "INFO  09:45:10,636 ProgressMeter -      I:11602244              0.0    10.5 m         1041.8 w       11.6%    90.8 m      80.3 m \n",
      "INFO  09:45:40,639 ProgressMeter -      I:12015240              0.0    11.0 m         1091.4 w       12.0%    91.8 m      80.8 m \n",
      "INFO  09:46:10,641 ProgressMeter -      I:12277688              0.0    11.5 m         1141.0 w       12.2%    93.9 m      82.4 m \n",
      "INFO  09:46:40,643 ProgressMeter -      I:12628273              0.0    12.0 m         1190.6 w       12.6%    95.3 m      83.3 m \n",
      "INFO  09:47:10,645 ProgressMeter -      I:12887270              0.0    12.5 m         1240.2 w       12.9%    97.3 m      84.8 m \n",
      "INFO  09:47:40,648 ProgressMeter -      I:13216943              0.0    13.0 m         1289.8 w       13.2%    98.6 m      85.6 m \n",
      "INFO  09:48:10,651 ProgressMeter -      I:13538415              0.0    13.5 m         1339.4 w       13.5%   100.0 m      86.5 m \n",
      "INFO  09:48:40,654 ProgressMeter -      I:14019024              0.0    14.0 m         1389.0 w       14.0%   100.2 m      86.2 m \n",
      "INFO  09:49:10,656 ProgressMeter -      I:14457757              0.0    14.5 m         1438.6 w       14.4%   100.6 m      86.1 m \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  09:49:40,658 ProgressMeter -      I:14979706              0.0    15.0 m         1488.2 w       14.9%   100.4 m      85.4 m \n",
      "INFO  09:50:10,660 ProgressMeter -      I:15065076              0.0    15.5 m         1537.8 w       15.0%   103.2 m      87.7 m \n",
      "INFO  09:50:40,662 ProgressMeter -      I:15066246              0.0    16.0 m         1587.4 w       15.0%   106.5 m      90.5 m \n",
      "INFO  09:51:10,664 ProgressMeter -      I:15066947              0.0    16.5 m         1637.0 w       15.0%   109.8 m      93.3 m \n",
      "INFO  09:51:40,667 ProgressMeter -       II:262002      1.5072434E7    17.0 m           67.0 s       15.3%   111.2 m      94.2 m \n",
      "INFO  09:52:10,669 ProgressMeter -       II:652570      1.5072434E7    17.5 m           69.0 s       15.7%   111.6 m      94.1 m \n",
      "INFO  09:52:40,671 ProgressMeter -       II:962259      1.5072434E7    18.0 m           71.0 s       16.0%   112.6 m      94.6 m \n",
      "INFO  09:53:10,673 ProgressMeter -      II:1059221      1.5072434E7    18.5 m           73.0 s       16.1%   115.0 m      96.5 m \n",
      "INFO  09:53:40,675 ProgressMeter -      II:1533675      1.5072434E7    19.0 m           75.0 s       16.6%   114.7 m      95.7 m \n",
      "INFO  09:54:10,678 ProgressMeter -      II:1613964      1.5072434E7    19.5 m           77.0 s       16.6%   117.2 m      97.7 m \n",
      "INFO  09:54:40,692 ProgressMeter -      II:1688827      1.5072434E7    20.0 m           79.0 s       16.7%   119.7 m      99.7 m \n",
      "INFO  09:55:10,695 ProgressMeter -      II:1793373      1.5072434E7    20.5 m           81.0 s       16.8%     2.0 h     101.4 m \n",
      "INFO  09:55:40,699 ProgressMeter -      II:1916390      1.5072434E7    21.0 m           83.0 s       16.9%     2.1 h     103.0 m \n",
      "INFO  09:56:10,702 ProgressMeter -      II:2085310      1.5072434E7    21.5 m           85.0 s       17.1%     2.1 h     104.2 m \n",
      "INFO  09:56:42,945 ProgressMeter -      II:2255836      1.5072434E7    22.0 m           87.0 s       17.3%     2.1 h     105.3 m \n",
      "INFO  09:57:12,951 ProgressMeter -      II:2347369      1.5072434E7    22.5 m           89.0 s       17.4%     2.2 h     107.0 m \n",
      "INFO  09:57:42,953 ProgressMeter -      II:2508942      1.5072434E7    23.0 m           91.0 s       17.5%     2.2 h     108.2 m \n",
      "INFO  09:58:12,955 ProgressMeter -      II:2655615      1.5072434E7    23.5 m           93.0 s       17.7%     2.2 h     109.4 m \n",
      "INFO  09:58:42,957 ProgressMeter -      II:2853414      1.5072434E7    24.0 m           95.0 s       17.9%     2.2 h     110.3 m \n",
      "INFO  09:59:12,960 ProgressMeter -      II:3169758      1.5072434E7    24.5 m           97.0 s       18.2%     2.2 h     110.2 m \n",
      "INFO  09:59:42,962 ProgressMeter -      II:3288314      1.5072434E7    25.0 m           99.0 s       18.3%     2.3 h     111.6 m \n",
      "INFO  10:00:12,964 ProgressMeter -      II:3469301      1.5072434E7    25.5 m          101.0 s       18.5%     2.3 h     112.4 m \n",
      "INFO  10:00:42,967 ProgressMeter -      II:3688926      1.5072434E7    26.0 m          103.0 s       18.7%     2.3 h     113.0 m \n",
      "INFO  10:01:12,969 ProgressMeter -      II:3785627      1.5072434E7    26.5 m          105.0 s       18.8%     2.3 h     114.4 m \n",
      "INFO  10:01:42,970 ProgressMeter -      II:3957039      1.5072434E7    27.0 m          107.0 s       19.0%     2.4 h     115.3 m \n",
      "INFO  10:02:12,972 ProgressMeter -      II:4392340      1.5072434E7    27.5 m          109.0 s       19.4%     2.4 h     114.2 m \n",
      "INFO  10:02:42,974 ProgressMeter -      II:4938000      1.5072434E7    28.0 m          111.0 s       20.0%     2.3 h     112.3 m \n",
      "INFO  10:03:12,977 ProgressMeter -      II:5571819      1.5072434E7    28.5 m          113.0 s       20.6%     2.3 h     109.9 m \n",
      "INFO  10:03:42,979 ProgressMeter -      II:6247362      1.5072434E7    29.0 m          115.0 s       21.3%     2.3 h     107.4 m \n",
      "INFO  10:04:12,982 ProgressMeter -      II:6891943      1.5072434E7    29.5 m          117.0 s       21.9%     2.2 h     105.2 m \n",
      "INFO  10:04:42,985 ProgressMeter -      II:7415167      1.5072434E7    30.0 m          119.0 s       22.4%     2.2 h     103.8 m \n",
      "INFO  10:05:12,987 ProgressMeter -      II:8014882      1.5072434E7    30.5 m            2.0 m       23.0%     2.2 h     102.0 m \n",
      "INFO  10:05:42,990 ProgressMeter -      II:8522113      1.5072434E7    31.0 m            2.1 m       23.5%     2.2 h     100.8 m \n",
      "INFO  10:06:12,992 ProgressMeter -      II:9162289      1.5072434E7    31.5 m            2.1 m       24.2%     2.2 h      98.9 m \n",
      "INFO  10:06:42,995 ProgressMeter -      II:9765475      1.5072434E7    32.0 m            2.1 m       24.8%     2.2 h      97.2 m \n",
      "INFO  10:07:12,998 ProgressMeter -     II:10489345      1.5072434E7    32.5 m            2.2 m       25.5%     2.1 h      95.0 m \n",
      "INFO  10:07:43,001 ProgressMeter -     II:11165308      1.5072434E7    33.0 m            2.2 m       26.2%     2.1 h      93.1 m \n",
      "INFO  10:08:13,003 ProgressMeter -     II:11719295      1.5072434E7    33.5 m            2.2 m       26.7%     2.1 h      91.9 m \n",
      "INFO  10:08:43,004 ProgressMeter -     II:12175178      1.5072434E7    34.0 m            2.3 m       27.2%     2.1 h      91.1 m \n",
      "INFO  10:09:13,007 ProgressMeter -     II:12511829      1.5072434E7    34.5 m            2.3 m       27.5%     2.1 h      90.9 m \n",
      "INFO  10:09:43,009 ProgressMeter -     II:12764371      1.5072434E7    35.0 m            2.3 m       27.8%     2.1 h      91.1 m \n",
      "INFO  10:10:13,012 ProgressMeter -     II:13203058      1.5072434E7    35.5 m            2.4 m       28.2%     2.1 h      90.4 m \n",
      "INFO  10:10:43,014 ProgressMeter -     II:13595353      1.5072434E7    36.0 m            2.4 m       28.6%     2.1 h      89.9 m \n",
      "INFO  10:11:13,017 ProgressMeter -     II:13806565      1.5072434E7    36.5 m            2.4 m       28.8%     2.1 h      90.3 m \n",
      "INFO  10:11:43,019 ProgressMeter -     II:14274990      1.5072434E7    37.0 m            2.5 m       29.3%     2.1 h      89.4 m \n",
      "INFO  10:12:15,342 ProgressMeter -     II:14803136      1.5072434E7    37.5 m            2.5 m       29.8%     2.1 h      88.4 m \n",
      "INFO  10:12:45,344 ProgressMeter -     II:15270809      1.5072434E7    38.0 m            2.5 m       30.3%     2.1 h      87.6 m \n",
      "INFO  10:13:15,346 ProgressMeter -       III:66582      3.0351855E7    38.5 m           76.0 s       30.3%     2.1 h      88.4 m \n",
      "INFO  10:13:45,349 ProgressMeter -      III:142866      3.0351855E7    39.0 m           77.0 s       30.4%     2.1 h      89.3 m \n",
      "INFO  10:14:15,351 ProgressMeter -      III:217275      3.0351855E7    39.5 m           78.0 s       30.5%     2.2 h      90.1 m \n",
      "INFO  10:14:45,354 ProgressMeter -      III:761836      3.0351855E7    40.0 m           79.0 s       31.0%     2.1 h      88.9 m \n",
      "INFO  10:15:15,358 ProgressMeter -      III:967402      3.0351855E7    40.5 m           80.0 s       31.2%     2.2 h      89.2 m \n",
      "INFO  10:15:45,362 ProgressMeter -     III:1275060      3.0351855E7    41.0 m           81.0 s       31.5%     2.2 h      89.0 m \n",
      "INFO  10:16:15,366 ProgressMeter -     III:1346652      3.0351855E7    41.5 m           82.0 s       31.6%     2.2 h      89.8 m \n",
      "INFO  10:16:45,369 ProgressMeter -     III:1469773      3.0351855E7    42.0 m           83.0 s       31.7%     2.2 h      90.4 m \n",
      "INFO  10:17:15,371 ProgressMeter -     III:1924775      3.0351855E7    42.5 m           84.0 s       32.2%     2.2 h      89.6 m \n",
      "INFO  10:17:45,376 ProgressMeter -     III:2441985      3.0351855E7    43.0 m           85.0 s       32.7%     2.2 h      88.5 m \n",
      "INFO  10:18:15,378 ProgressMeter -     III:2926086      3.0351855E7    43.5 m           85.0 s       33.2%     2.2 h      87.6 m \n",
      "INFO  10:18:45,379 ProgressMeter -     III:3438319      3.0351855E7    44.0 m           86.0 s       33.7%     2.2 h      86.6 m \n",
      "INFO  10:19:15,382 ProgressMeter -     III:4047560      3.0351855E7    44.5 m           87.0 s       34.3%     2.2 h      85.2 m \n",
      "INFO  10:19:45,384 ProgressMeter -     III:4865682      3.0351855E7    45.0 m           88.0 s       35.1%     2.1 h      83.1 m \n",
      "INFO  10:20:15,388 ProgressMeter -     III:5579252      3.0351855E7    45.5 m           89.0 s       35.8%     2.1 h      81.5 m \n",
      "INFO  10:20:45,390 ProgressMeter -     III:6330376      3.0351855E7    46.0 m           90.0 s       36.6%     2.1 h      79.8 m \n",
      "INFO  10:21:15,392 ProgressMeter -     III:7003350      3.0351855E7    46.5 m           91.0 s       37.2%     2.1 h      78.3 m \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  10:21:45,402 ProgressMeter -     III:7604431      3.0351855E7    47.0 m           92.0 s       37.8%     2.1 h      77.2 m \n",
      "INFO  10:22:15,404 ProgressMeter -     III:8129755      3.0351855E7    47.5 m           93.0 s       38.4%     2.1 h      76.3 m \n",
      "INFO  10:22:45,407 ProgressMeter -     III:8852946      3.0351855E7    48.0 m           94.0 s       39.1%     2.0 h      74.8 m \n",
      "INFO  10:23:15,409 ProgressMeter -     III:9605260      3.0351855E7    48.5 m           95.0 s       39.8%     2.0 h      73.2 m \n",
      "INFO  10:23:45,410 ProgressMeter -    III:10311882      3.0351855E7    49.0 m           96.0 s       40.5%     2.0 h      71.8 m \n",
      "INFO  10:24:15,412 ProgressMeter -    III:10923163      3.0351855E7    49.5 m           97.0 s       41.2%     2.0 h      70.8 m \n",
      "INFO  10:24:45,413 ProgressMeter -    III:11569274      3.0351855E7    50.0 m           98.0 s       41.8%   119.6 m      69.6 m \n",
      "INFO  10:25:15,415 ProgressMeter -    III:11962370      3.0351855E7    50.5 m           99.0 s       42.2%   119.7 m      69.2 m \n",
      "INFO  10:25:45,417 ProgressMeter -    III:12375065      3.0351855E7    51.0 m          100.0 s       42.6%   119.7 m      68.7 m \n",
      "INFO  10:26:15,419 ProgressMeter -    III:12703001      3.0351855E7    51.5 m          101.0 s       42.9%   120.0 m      68.5 m \n",
      "INFO  10:26:45,420 ProgressMeter -    III:13146365      3.0351855E7    52.0 m          102.0 s       43.4%   119.9 m      67.9 m \n",
      "INFO  10:27:15,422 ProgressMeter -       IV:185486      4.4135656E7    52.5 m           71.0 s       44.2%   118.8 m      66.3 m \n",
      "INFO  10:27:47,026 ProgressMeter -       IV:989403      4.4135656E7    53.0 m           72.0 s       45.0%   117.8 m      64.8 m \n",
      "INFO  10:28:17,028 ProgressMeter -      IV:1453918      4.4135656E7    53.5 m           72.0 s       45.5%   117.7 m      64.2 m \n",
      "INFO  10:28:47,031 ProgressMeter -      IV:1837988      4.4135656E7    54.0 m           73.0 s       45.8%   117.8 m      63.8 m \n",
      "INFO  10:29:17,032 ProgressMeter -      IV:2334618      4.4135656E7    54.5 m           74.0 s       46.3%   117.6 m      63.1 m \n",
      "INFO  10:29:47,034 ProgressMeter -      IV:2661277      4.4135656E7    55.0 m           74.0 s       46.7%   117.9 m      62.9 m \n",
      "INFO  10:30:17,036 ProgressMeter -      IV:2948251      4.4135656E7    55.5 m           75.0 s       46.9%   118.2 m      62.7 m \n",
      "INFO  10:30:47,038 ProgressMeter -      IV:3422948      4.4135656E7    56.0 m           76.0 s       47.4%   118.1 m      62.1 m \n",
      "INFO  10:31:17,040 ProgressMeter -      IV:3923049      4.4135656E7    56.5 m           76.0 s       47.9%   117.9 m      61.4 m \n",
      "INFO  10:31:47,057 ProgressMeter -      IV:3985104      4.4135656E7    57.0 m           77.0 s       48.0%   118.8 m      61.8 m \n",
      "INFO  10:32:17,059 ProgressMeter -      IV:4493613      4.4135656E7    57.5 m           78.0 s       48.5%   118.6 m      61.1 m \n",
      "INFO  10:32:47,061 ProgressMeter -      IV:5374922      4.4135656E7    58.0 m           78.0 s       49.4%   117.5 m      59.5 m \n",
      "INFO  10:33:17,064 ProgressMeter -      IV:6107465      4.4135656E7    58.5 m           79.0 s       50.1%   116.8 m      58.3 m \n",
      "INFO  10:33:47,066 ProgressMeter -      IV:6320519      4.4135656E7    59.0 m           80.0 s       50.3%   117.3 m      58.3 m \n",
      "INFO  10:34:17,069 ProgressMeter -      IV:6969370      4.4135656E7    59.5 m           80.0 s       51.0%   116.8 m      57.3 m \n",
      "INFO  10:34:47,070 ProgressMeter -      IV:7537686      4.4135656E7    60.0 m           81.0 s       51.5%   116.4 m      56.4 m \n",
      "INFO  10:35:17,073 ProgressMeter -      IV:8246182      4.4135656E7    60.5 m           82.0 s       52.2%   115.8 m      55.3 m \n",
      "INFO  10:35:47,076 ProgressMeter -      IV:8579646      4.4135656E7    61.0 m           82.0 s       52.6%   116.0 m      55.0 m \n",
      "INFO  10:36:17,079 ProgressMeter -      IV:9188178      4.4135656E7    61.5 m           83.0 s       53.2%   115.7 m      54.2 m \n",
      "INFO  10:36:47,081 ProgressMeter -      IV:9940123      4.4135656E7    62.0 m           84.0 s       53.9%   115.0 m      53.0 m \n",
      "INFO  10:37:17,083 ProgressMeter -     IV:10669195      4.4135656E7    62.5 m           84.0 s       54.6%   114.4 m      51.9 m \n",
      "INFO  10:37:47,084 ProgressMeter -     IV:11361874      4.4135656E7    63.0 m           85.0 s       55.3%   113.8 m      50.8 m \n",
      "INFO  10:38:17,086 ProgressMeter -     IV:12101822      4.4135656E7    63.5 m           86.0 s       56.1%   113.2 m      49.7 m \n",
      "INFO  10:38:47,087 ProgressMeter -     IV:12832694      4.4135656E7    64.0 m           87.0 s       56.8%   112.7 m      48.7 m \n",
      "INFO  10:39:17,089 ProgressMeter -     IV:13313293      4.4135656E7    64.5 m           87.0 s       57.3%   112.6 m      48.1 m \n",
      "INFO  10:39:47,092 ProgressMeter -     IV:13764914      4.4135656E7    65.0 m           88.0 s       57.7%   112.6 m      47.6 m \n",
      "INFO  10:40:17,094 ProgressMeter -     IV:14209447      4.4135656E7    65.5 m           89.0 s       58.2%   112.6 m      47.1 m \n",
      "INFO  10:40:47,097 ProgressMeter -     IV:14578826      4.4135656E7    66.0 m           89.0 s       58.5%   112.7 m      46.7 m \n",
      "INFO  10:41:17,099 ProgressMeter -     IV:15053034      4.4135656E7    66.5 m           90.0 s       59.0%   112.7 m      46.2 m \n",
      "INFO  10:41:47,100 ProgressMeter -     IV:15431040      4.4135656E7    67.0 m           91.0 s       59.4%   112.8 m      45.8 m \n",
      "INFO  10:42:17,103 ProgressMeter -     IV:15769730      4.4135656E7    67.5 m           91.0 s       59.7%   113.0 m      45.5 m \n",
      "INFO  10:42:47,105 ProgressMeter -     IV:16220340      4.4135656E7    68.0 m           92.0 s       60.2%   113.0 m      45.0 m \n",
      "INFO  10:43:19,654 ProgressMeter -     IV:16490872      4.4135656E7    68.5 m           93.0 s       60.5%   113.3 m      44.8 m \n",
      "INFO  10:43:49,657 ProgressMeter -     IV:16859207      4.4135656E7    69.0 m           93.0 s       60.8%   113.4 m      44.4 m \n",
      "INFO  10:44:19,659 ProgressMeter -     IV:17472229      4.4135656E7    69.5 m           94.0 s       61.4%   113.1 m      43.6 m \n",
      "INFO  10:44:49,661 ProgressMeter -        V:520769      6.1629485E7    70.0 m           68.0 s       62.0%   113.0 m      43.0 m \n",
      "INFO  10:45:19,664 ProgressMeter -        V:563417      6.1629485E7    70.5 m           68.0 s       62.0%   113.7 m      43.2 m \n",
      "INFO  10:45:49,666 ProgressMeter -        V:677153      6.1629485E7    71.0 m           69.0 s       62.1%   114.3 m      43.3 m \n",
      "INFO  10:46:19,668 ProgressMeter -        V:893193      6.1629485E7    71.5 m           69.0 s       62.3%   114.7 m      43.2 m \n",
      "INFO  10:46:49,670 ProgressMeter -       V:1104874      6.1629485E7    72.0 m           70.0 s       62.6%   115.1 m      43.1 m \n",
      "INFO  10:47:19,673 ProgressMeter -       V:1474363      6.1629485E7    72.5 m           70.0 s       62.9%   115.2 m      42.7 m \n",
      "INFO  10:47:49,676 ProgressMeter -       V:1827019      6.1629485E7    73.0 m           71.0 s       63.3%   115.4 m      42.4 m \n",
      "INFO  10:48:19,678 ProgressMeter -       V:2192521      6.1629485E7    73.5 m           71.0 s       63.6%   115.5 m      42.0 m \n",
      "INFO  10:48:49,681 ProgressMeter -       V:2308233      6.1629485E7    74.0 m           72.0 s       63.8%   116.1 m      42.1 m \n",
      "INFO  10:49:19,683 ProgressMeter -       V:2744932      6.1629485E7    74.5 m           72.0 s       64.2%   116.1 m      41.6 m \n",
      "INFO  10:49:49,685 ProgressMeter -       V:2900694      6.1629485E7    75.0 m           73.0 s       64.3%   116.6 m      41.6 m \n",
      "INFO  10:50:19,688 ProgressMeter -       V:3190178      6.1629485E7    75.5 m           73.0 s       64.6%   116.8 m      41.3 m \n",
      "INFO  10:50:49,690 ProgressMeter -       V:3410760      6.1629485E7    76.0 m           73.0 s       64.9%   117.2 m      41.2 m \n",
      "INFO  10:51:19,693 ProgressMeter -       V:3619170      6.1629485E7    76.5 m           74.0 s       65.1%   117.6 m      41.1 m \n",
      "INFO  10:51:49,695 ProgressMeter -       V:3860703      6.1629485E7    77.0 m           74.0 s       65.3%   117.9 m      40.9 m \n",
      "INFO  10:52:19,697 ProgressMeter -       V:3954075      6.1629485E7    77.5 m           75.0 s       65.4%   118.5 m      41.0 m \n",
      "INFO  10:52:49,699 ProgressMeter -       V:4159529      6.1629485E7    78.0 m           75.0 s       65.6%   118.9 m      40.9 m \n",
      "INFO  10:53:19,701 ProgressMeter -       V:4838348      6.1629485E7    78.5 m           76.0 s       66.3%   118.4 m      39.9 m \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  10:53:49,704 ProgressMeter -       V:5249995      6.1629485E7    79.0 m           76.0 s       66.7%   118.5 m      39.5 m \n",
      "INFO  10:54:19,706 ProgressMeter -       V:5839293      6.1629485E7    79.5 m           77.0 s       67.3%   118.2 m      38.7 m \n",
      "INFO  10:54:49,708 ProgressMeter -       V:6379258      6.1629485E7    80.0 m           77.0 s       67.8%   118.0 m      38.0 m \n",
      "INFO  10:55:19,711 ProgressMeter -       V:6788518      6.1629485E7    80.5 m           78.0 s       68.2%   118.0 m      37.5 m \n",
      "INFO  10:55:49,714 ProgressMeter -       V:7326831      6.1629485E7    81.0 m           78.0 s       68.8%   117.8 m      36.8 m \n",
      "INFO  10:56:19,716 ProgressMeter -       V:7382923      6.1629485E7    81.5 m           79.0 s       68.8%   118.4 m      36.9 m \n",
      "INFO  10:56:49,718 ProgressMeter -       V:7611090      6.1629485E7    82.0 m           79.0 s       69.0%   118.8 m      36.8 m \n",
      "INFO  10:57:19,721 ProgressMeter -       V:7643292      6.1629485E7    82.5 m           80.0 s       69.1%   119.4 m      36.9 m \n",
      "INFO  10:57:49,723 ProgressMeter -       V:7814247      6.1629485E7    83.0 m           80.0 s       69.2%   119.9 m      36.9 m \n",
      "INFO  10:58:19,725 ProgressMeter -       V:8285294      6.1629485E7    83.5 m           81.0 s       69.7%   119.8 m      36.3 m \n",
      "INFO  10:58:49,727 ProgressMeter -       V:9103520      6.1629485E7    84.0 m           81.0 s       70.5%   119.1 m      35.1 m \n",
      "INFO  10:59:19,730 ProgressMeter -       V:9809781      6.1629485E7    84.5 m           82.0 s       71.2%   118.6 m      34.1 m \n",
      "INFO  10:59:49,732 ProgressMeter -      V:10575549      6.1629485E7    85.0 m           82.0 s       72.0%   118.1 m      33.1 m \n",
      "INFO  11:00:19,734 ProgressMeter -      V:11197310      6.1629485E7    85.5 m           83.0 s       72.6%   117.7 m      32.2 m \n",
      "INFO  11:00:49,736 ProgressMeter -      V:11814516      6.1629485E7    86.0 m           83.0 s       73.2%   117.4 m      31.4 m \n",
      "INFO  11:01:19,738 ProgressMeter -      V:12399105      6.1629485E7    86.5 m           84.0 s       73.8%   117.2 m      30.7 m \n",
      "INFO  11:01:49,741 ProgressMeter -      V:12851358      6.1629485E7    87.0 m           84.0 s       74.3%   117.1 m      30.1 m \n",
      "INFO  11:02:19,743 ProgressMeter -      V:13555522      6.1629485E7    87.5 m           85.0 s       75.0%   116.7 m      29.2 m \n",
      "INFO  11:02:49,745 ProgressMeter -      V:14246853      6.1629485E7    88.0 m           85.0 s       75.7%   116.3 m      28.3 m \n",
      "INFO  11:03:19,747 ProgressMeter -      V:14855834      6.1629485E7    88.5 m           86.0 s       76.3%   116.0 m      27.5 m \n",
      "INFO  11:03:49,754 ProgressMeter -      V:15357951      6.1629485E7    89.0 m           86.0 s       76.8%   115.9 m      26.9 m \n",
      "INFO  11:04:19,756 ProgressMeter -      V:15673605      6.1629485E7    89.5 m           87.0 s       77.1%   116.1 m      26.6 m \n",
      "INFO  11:04:49,758 ProgressMeter -      V:15755812      6.1629485E7    90.0 m           87.0 s       77.2%   116.6 m      26.6 m \n",
      "INFO  11:05:19,762 ProgressMeter -      V:15912744      6.1629485E7    90.5 m           88.0 s       77.3%   117.0 m      26.5 m \n",
      "INFO  11:05:49,764 ProgressMeter -      V:15961604      6.1629485E7    91.0 m           88.0 s       77.4%   117.6 m      26.6 m \n",
      "INFO  11:06:19,766 ProgressMeter -      V:16139308      6.1629485E7    91.5 m           89.0 s       77.5%   118.0 m      26.5 m \n",
      "INFO  11:06:49,767 ProgressMeter -      V:16216634      6.1629485E7    92.0 m           89.0 s       77.6%   118.5 m      26.5 m \n",
      "INFO  11:07:23,195 ProgressMeter -      V:16376878      6.1629485E7    92.5 m           90.0 s       77.8%   118.9 m      26.4 m \n",
      "INFO  11:07:53,197 ProgressMeter -      V:16490010      6.1629485E7    93.0 m           90.0 s       77.9%   119.4 m      26.4 m \n",
      "INFO  11:08:23,199 ProgressMeter -      V:16607228      6.1629485E7    93.5 m           91.0 s       78.0%   119.9 m      26.4 m \n",
      "INFO  11:08:53,202 ProgressMeter -      V:16862849      6.1629485E7    94.0 m           91.0 s       78.3%     2.0 h      26.1 m \n",
      "INFO  11:09:23,204 ProgressMeter -      V:16988532      6.1629485E7    94.5 m           92.0 s       78.4%     2.0 h      26.0 m \n",
      "INFO  11:09:53,206 ProgressMeter -      V:17225523      6.1629485E7    95.0 m           92.0 s       78.6%     2.0 h      25.8 m \n",
      "INFO  11:10:23,209 ProgressMeter -      V:17329110      6.1629485E7    95.5 m           92.0 s       78.7%     2.0 h      25.8 m \n",
      "INFO  11:10:53,211 ProgressMeter -      V:17423047      6.1629485E7    96.0 m           93.0 s       78.8%     2.0 h      25.8 m \n",
      "INFO  11:11:23,214 ProgressMeter -      V:17554914      6.1629485E7    96.5 m           93.0 s       79.0%     2.0 h      25.7 m \n",
      "INFO  11:11:53,216 ProgressMeter -      V:17674097      6.1629485E7    97.0 m           94.0 s       79.1%     2.0 h      25.7 m \n",
      "INFO  11:12:23,219 ProgressMeter -      V:17883379      6.1629485E7    97.5 m           94.0 s       79.3%     2.0 h      25.5 m \n",
      "INFO  11:12:53,221 ProgressMeter -      V:18135939      6.1629485E7    98.0 m           95.0 s       79.5%     2.1 h      25.2 m \n",
      "INFO  11:13:23,224 ProgressMeter -      V:18284338      6.1629485E7    98.5 m           95.0 s       79.7%     2.1 h      25.1 m \n",
      "INFO  11:13:53,226 ProgressMeter -      V:18652263      6.1629485E7    99.0 m           96.0 s       80.1%     2.1 h      24.7 m \n",
      "INFO  11:14:23,229 ProgressMeter -      V:18950480      6.1629485E7    99.5 m           96.0 s       80.3%     2.1 h      24.3 m \n",
      "INFO  11:14:53,231 ProgressMeter -      V:19237027      6.1629485E7   100.0 m           97.0 s       80.6%     2.1 h      24.0 m \n",
      "INFO  11:15:23,234 ProgressMeter -      V:19304077      6.1629485E7   100.5 m           97.0 s       80.7%     2.1 h      24.0 m \n",
      "INFO  11:15:53,236 ProgressMeter -      V:19371916      6.1629485E7   101.0 m           98.0 s       80.8%     2.1 h      24.0 m \n",
      "INFO  11:16:23,238 ProgressMeter -      V:19513770      6.1629485E7   101.5 m           98.0 s       80.9%     2.1 h      23.9 m \n",
      "INFO  11:16:53,241 ProgressMeter -      V:19847974      6.1629485E7   102.0 m           99.0 s       81.2%     2.1 h      23.5 m \n",
      "INFO  11:17:23,243 ProgressMeter -      V:20140695      6.1629485E7   102.5 m           99.0 s       81.5%     2.1 h      23.2 m \n",
      "INFO  11:17:53,246 ProgressMeter -      V:20391307      6.1629485E7   103.0 m          100.0 s       81.8%     2.1 h      22.9 m \n",
      "INFO  11:18:23,248 ProgressMeter -      V:20547755      6.1629485E7   103.5 m          100.0 s       81.9%     2.1 h      22.8 m \n",
      "INFO  11:18:53,250 ProgressMeter -      V:20609449      6.1629485E7   104.0 m          101.0 s       82.0%     2.1 h      22.8 m \n",
      "INFO  11:19:23,253 ProgressMeter -         X:74762      8.2553665E7   104.5 m           75.0 s       82.4%     2.1 h      22.3 m \n",
      "INFO  11:19:53,255 ProgressMeter -        X:695756      8.2553665E7   105.0 m           76.0 s       83.0%     2.1 h      21.5 m \n",
      "INFO  11:20:23,257 ProgressMeter -       X:1279669      8.2553665E7   105.5 m           76.0 s       83.6%     2.1 h      20.7 m \n",
      "INFO  11:20:53,260 ProgressMeter -       X:1566343      8.2553665E7   106.0 m           77.0 s       83.9%     2.1 h      20.4 m \n",
      "INFO  11:21:23,262 ProgressMeter -       X:1972137      8.2553665E7   106.5 m           77.0 s       84.3%     2.1 h      19.9 m \n",
      "INFO  11:21:53,266 ProgressMeter -       X:2560581      8.2553665E7   107.0 m           77.0 s       84.9%     2.1 h      19.1 m \n",
      "INFO  11:22:23,268 ProgressMeter -       X:3139250      8.2553665E7   107.5 m           78.0 s       85.4%     2.1 h      18.3 m \n",
      "INFO  11:22:53,270 ProgressMeter -       X:3768144      8.2553665E7   108.0 m           78.0 s       86.1%     2.1 h      17.5 m \n",
      "INFO  11:23:23,272 ProgressMeter -       X:4396510      8.2553665E7   108.5 m           78.0 s       86.7%     2.1 h      16.6 m \n",
      "INFO  11:23:53,274 ProgressMeter -       X:4992661      8.2553665E7   109.0 m           79.0 s       87.3%     2.1 h      15.9 m \n",
      "INFO  11:24:23,275 ProgressMeter -       X:5526203      8.2553665E7   109.5 m           79.0 s       87.8%     2.1 h      15.2 m \n",
      "INFO  11:24:53,277 ProgressMeter -       X:6209910      8.2553665E7   110.0 m           79.0 s       88.5%     2.1 h      14.3 m \n",
      "INFO  11:25:23,279 ProgressMeter -       X:6835409      8.2553665E7   110.5 m           80.0 s       89.1%     2.1 h      13.5 m \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  11:25:53,282 ProgressMeter -       X:7283496      8.2553665E7   111.0 m           80.0 s       89.6%     2.1 h      12.9 m \n",
      "INFO  11:26:23,284 ProgressMeter -       X:7921419      8.2553665E7   111.5 m           81.0 s       90.2%     2.1 h      12.1 m \n",
      "INFO  11:26:53,286 ProgressMeter -       X:8545386      8.2553665E7   112.0 m           81.0 s       90.8%     2.1 h      11.3 m \n",
      "INFO  11:27:23,288 ProgressMeter -       X:9175674      8.2553665E7   112.5 m           81.0 s       91.5%     2.0 h      10.5 m \n",
      "INFO  11:27:53,290 ProgressMeter -       X:9818274      8.2553665E7   113.0 m           82.0 s       92.1%     2.0 h       9.7 m \n",
      "INFO  11:28:23,292 ProgressMeter -      X:10487150      8.2553665E7   113.5 m           82.0 s       92.8%     2.0 h       8.8 m \n",
      "INFO  11:28:53,294 ProgressMeter -      X:11132064      8.2553665E7   114.0 m           82.0 s       93.4%     2.0 h       8.0 m \n",
      "INFO  11:29:23,297 ProgressMeter -      X:11773262      8.2553665E7   114.5 m           83.0 s       94.1%     2.0 h       7.2 m \n",
      "INFO  11:29:53,304 ProgressMeter -      X:12373970      8.2553665E7   115.0 m           83.0 s       94.7%     2.0 h       6.5 m \n",
      "INFO  11:30:23,306 ProgressMeter -      X:12982289      8.2553665E7   115.5 m           83.0 s       95.3%     2.0 h       5.7 m \n",
      "INFO  11:30:56,742 ProgressMeter -      X:13527050      8.2553665E7   116.0 m           84.0 s       95.8%     2.0 h       5.1 m \n",
      "INFO  11:31:26,744 ProgressMeter -      X:14170289      8.2553665E7   116.5 m           84.0 s       96.4%     2.0 h       4.3 m \n",
      "INFO  11:31:56,747 ProgressMeter -      X:14471513      8.2553665E7   117.0 m           85.0 s       96.7%     2.0 h       3.9 m \n",
      "INFO  11:32:26,749 ProgressMeter -      X:14902362      8.2553665E7   117.5 m           85.0 s       97.2%     2.0 h       3.4 m \n",
      "INFO  11:32:56,750 ProgressMeter -      X:15359101      8.2553665E7   118.0 m           85.0 s       97.6%     2.0 h       2.9 m \n",
      "INFO  11:33:26,753 ProgressMeter -      X:15840969      8.2553665E7   118.5 m           86.0 s       98.1%     2.0 h       2.3 m \n",
      "INFO  11:33:56,755 ProgressMeter -      X:16347335      8.2553665E7   119.0 m           86.0 s       98.6%     2.0 h     100.0 s \n",
      "INFO  11:34:26,757 ProgressMeter -      X:17037240      8.2553665E7   119.5 m           86.0 s       99.3%     2.0 h      50.0 s \n",
      "INFO  11:34:56,759 ProgressMeter -      X:17556138      8.2553665E7   120.0 m           87.0 s       99.8%     2.0 h      12.0 s \n",
      "INFO  11:35:26,761 ProgressMeter -      MtDNA:9707     1.00272607E8     2.0 h           72.0 s      100.0%     2.0 h       0.0 s \n",
      "INFO  11:35:29,004 VectorLoglessPairHMM - Time spent in setup for JNI call : 3.881168494 \n",
      "INFO  11:35:29,005 PairHMM - Total compute time in PairHMM computeLikelihoods() : 1250.1589546270002 \n",
      "INFO  11:35:29,005 HaplotypeCaller - Ran local assembly on 657229 active regions \n",
      "INFO  11:35:29,015 ProgressMeter -            done     1.00286401E8     2.0 h           72.0 s      100.0%     2.0 h       0.0 s \n",
      "INFO  11:35:29,016 ProgressMeter - Total runtime 7232.86 secs, 120.55 min, 2.01 hours \n",
      "INFO  11:35:29,018 MicroScheduler - 404689 reads were filtered out during the traversal out of approximately 34819284 total reads (1.16%) \n",
      "INFO  11:35:29,020 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  11:35:29,022 MicroScheduler -   -> 0 reads (0.00% of total) failing DuplicateReadFilter \n",
      "INFO  11:35:29,023 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  11:35:29,024 MicroScheduler -   -> 385649 reads (1.11% of total) failing HCMappingQualityFilter \n",
      "INFO  11:35:29,025 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  11:35:29,026 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  11:35:29,027 MicroScheduler -   -> 19040 reads (0.05% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  11:35:29,027 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were 4 WARN messages, the first 4 are repeated below.\n",
      "WARN  09:34:38,857 InbreedingCoeff - Annotation will not be calculated. InbreedingCoeff requires at least 10 unrelated samples. \n",
      "WARN  09:34:40,174 HaplotypeScore - Annotation will not be calculated, must be called from UnifiedGenotyper, not HaplotypeCaller \n",
      "WARN  09:34:40,429 AnnotationUtils - Annotation will not be calculated, genotype is not called \n",
      "WARN  09:34:40,429 AnnotationUtils - Annotation will not be calculated, genotype is not called \n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/gatk-3.8.1.0/GenomeAnalysisTK.jar -T HaplotypeCaller --emitRefConfidence GVCF -R {ReferenceGenome} -l INFO -I {path}/analysis/Alignments/{SampleName}.realigned.bam -o {SampleName}.g.vcf -mbq {minBaseQforVcf} -mmq {minMaqQforVcf} -stand_call_conf {minVarCall}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  12:03:19,468 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  12:03:19,470 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.8-1-0-gf15c1c3ef, Compiled 2018/02/19 05:43:50 \n",
      "INFO  12:03:19,470 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  12:03:19,470 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  12:03:19,470 HelpFormatter - [Sun Jun 20 12:03:19 AST 2021] Executing on Linux 4.4.0-210-generic amd64 \n",
      "INFO  12:03:19,470 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10 \n",
      "INFO  12:03:19,473 HelpFormatter - Program Args: -T GenotypeGVCFs -R /home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/data/Caenorhabditis_elegans.WBcel235.99.softmasked.fa -nt 20 -l INFO -V CB4856.g.vcf -dt none -o CB4856.HC.vcf \n",
      "INFO  12:03:19,478 HelpFormatter - Executing as velazqam@kw60530 on Linux 4.4.0-210-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_292-8u292-b10-0ubuntu1~16.04.1-b10. \n",
      "INFO  12:03:19,478 HelpFormatter - Date/Time: 2021/06/20 12:03:19 \n",
      "INFO  12:03:19,478 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  12:03:19,479 HelpFormatter - ------------------------------------------------------------------------------------ \n",
      "INFO  12:03:19,500 NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/home/velazqam/Documents/Projects/Jupyter/MappingBySequencing/Tutorial/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar!/com/intel/gkl/native/libgkl_compression.so \n",
      "INFO  12:03:19,528 GenomeAnalysisEngine - Deflater: IntelDeflater \n",
      "INFO  12:03:19,529 GenomeAnalysisEngine - Inflater: IntelInflater \n",
      "INFO  12:03:19,529 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  12:03:19,873 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  12:03:19,940 MicroScheduler - Running the GATK in parallel mode with 20 total threads, 1 CPU thread(s) for each of 20 data thread(s), of 32 processors available on this machine \n",
      "INFO  12:03:20,011 GenomeAnalysisEngine - Preparing for traversal \n",
      "INFO  12:03:20,013 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  12:03:20,013 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  12:03:20,013 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  12:03:20,013 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "WARN  12:03:20,098 StrandBiasTest - StrandBiasBySample annotation exists in input VCF header. Attempting to use StrandBiasBySample values to calculate strand bias annotation values. If no sample has the SB genotype annotation, annotation may still fail. \n",
      "WARN  12:03:20,099 InbreedingCoeff - Annotation will not be calculated. InbreedingCoeff requires at least 10 unrelated samples. \n",
      "WARN  12:03:20,099 StrandBiasTest - StrandBiasBySample annotation exists in input VCF header. Attempting to use StrandBiasBySample values to calculate strand bias annotation values. If no sample has the SB genotype annotation, annotation may still fail. \n",
      "INFO  12:03:20,100 GenotypeGVCFs - Notice that the -ploidy parameter is ignored in GenotypeGVCFs tool as this is automatically determined by the input variant files \n",
      "WARN  12:03:20,328 HaplotypeScore - Annotation will not be calculated, must be called from UnifiedGenotyper, not GenotypeGVCFs \n",
      "INFO  12:03:50,020 ProgressMeter -       V:3032801   4.6629485E7    30.0 s       0.0 s       64.5%    46.0 s      16.0 s \n",
      "INFO  12:04:12,590 ProgressMeter -            done   1.00286401E8    50.0 s       0.0 s      100.0%    50.0 s       0.0 s \n",
      "INFO  12:04:12,590 ProgressMeter - Total runtime 50.56 secs, 0.84 min, 0.01 hours \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were 4 WARN messages, the first 4 are repeated below.\n",
      "WARN  12:03:20,098 StrandBiasTest - StrandBiasBySample annotation exists in input VCF header. Attempting to use StrandBiasBySample values to calculate strand bias annotation values. If no sample has the SB genotype annotation, annotation may still fail. \n",
      "WARN  12:03:20,099 InbreedingCoeff - Annotation will not be calculated. InbreedingCoeff requires at least 10 unrelated samples. \n",
      "WARN  12:03:20,099 StrandBiasTest - StrandBiasBySample annotation exists in input VCF header. Attempting to use StrandBiasBySample values to calculate strand bias annotation values. If no sample has the SB genotype annotation, annotation may still fail. \n",
      "WARN  12:03:20,328 HaplotypeScore - Annotation will not be calculated, must be called from UnifiedGenotyper, not GenotypeGVCFs \n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!java -Xmx{ramG}g -jar {softpath}/gatk-3.8.1.0/GenomeAnalysisTK.jar -T GenotypeGVCFs -R {ReferenceGenome} -nt {Ncpu} -l INFO -V {SampleName}.g.vcf -dt none -o {SampleName}.HC.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now SnpEff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Move to analysis\n",
    "os.chdir(path + '/analysis/')\n",
    "##Create directory to place data\n",
    "!mkdir -p VariantPrediction\n",
    "##Move to that directory\n",
    "os.chdir(path + '/analysis/VariantPrediction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Produce Annotations\n",
    "!java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/snpEff.jar {SnpEffGen} -stats {SampleName}.HC.html {path}/analysis/VariantCalling/{SampleName}.HC.vcf > {SampleName}.HC.ann.vcf\n",
    "!java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/snpEff.jar {SnpEffGen} -stats {SampleName}.UG.html {path}/analysis/VariantCalling/{SampleName}.UG.vcf > {SampleName}.UG.ann.vcf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat {SampleName}.HC.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - -s \",\" -e \".\" CHROM POS REF ALT QUAL GEN[{SampleName}].GT GEN[{SampleName}].DP GEN[*].AD \"ANN[*].GENEID\" \"ANN[*].ALLELE\" \"ANN[*].EFFECT\" \"ANN[*].IMPACT\" \"ANN[*].GENE\" \"ANN[*].FEATURE\" \"ANN[*].FEATUREID\" \"ANN[*].BIOTYPE\" \"ANN[*].RANK\" \"ANN[*].HGVS_C\" \"ANN[*].HGVS_P\" \"ANN[*].CDNA_POS\" \"ANN[*].CDNA_LEN\" \"ANN[*].CDS_POS\" \"ANN[*].CDS_LEN\" \"ANN[*].AA_POS\" \"ANN[*].AA_LEN\" \"ANN[*].DISTANCE\" \"LOF[*].GENE\" \"LOF[*].GENEID\" \"NMD[*].GENE\" \"NMD[*].GENEID\" > {SampleName}.HC.ann.txt\n",
    "!cat {SampleName}.UG.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - -s \",\" -e \".\" CHROM POS REF ALT QUAL GEN[{SampleName}].GT GEN[{SampleName}].DP GEN[*].AD \"ANN[*].GENEID\" \"ANN[*].ALLELE\" \"ANN[*].EFFECT\" \"ANN[*].IMPACT\" \"ANN[*].GENE\" \"ANN[*].FEATURE\" \"ANN[*].FEATUREID\" \"ANN[*].BIOTYPE\" \"ANN[*].RANK\" \"ANN[*].HGVS_C\" \"ANN[*].HGVS_P\" \"ANN[*].CDNA_POS\" \"ANN[*].CDNA_LEN\" \"ANN[*].CDS_POS\" \"ANN[*].CDS_LEN\" \"ANN[*].AA_POS\" \"ANN[*].AA_LEN\" \"ANN[*].DISTANCE\" \"LOF[*].GENE\" \"LOF[*].GENEID\" \"NMD[*].GENE\" \"NMD[*].GENEID\" > {SampleName}.UG.ann.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## other approach\n",
    "!echo \"#High impact mutations\" > {SampleName}.HC.ann.sort.txt\n",
    "!cat {SampleName}.HC.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar filter \"ANN[0].IMPACT has 'HIGH'\" | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - CHROM POS REF ALT \"ANN[*].GENE\" \"ANN[*].FEATUREID\" \"ANN[*].EFFECT\" \"LOF[*].GENE\" >> {SampleName}.HC.ann.sort.txt\n",
    "!echo \"#Moderate impact mutations\" >> {SampleName}.HC.ann.sort.txt\n",
    "!cat {SampleName}.HC.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar filter \"ANN[0].IMPACT has 'MODERATE'\" | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - CHROM POS REF ALT \"ANN[*].GENE\" \"ANN[*].FEATUREID\" \"ANN[*].EFFECT\" \"LOF[*].GENE\" >> {SampleName}.HC.ann.sort.txt\n",
    "!echo \"#Low impact mutations\" >> {SampleName}.HC.ann.sort.txt\n",
    "!cat {SampleName}.HC.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar filter \"ANN[0].IMPACT has 'LOW'\" | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - CHROM POS REF ALT \"ANN[*].GENE\" \"ANN[*].FEATUREID\" \"ANN[*].EFFECT\" \"LOF[*].GENE\" >> {SampleName}.HC.ann.sort.txt\n",
    "\n",
    "!echo \"#High impact mutations\" > {SampleName}.UG.ann.sort.txt\n",
    "!cat {SampleName}.UG.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar filter \"ANN[0].IMPACT has 'HIGH'\" | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - CHROM POS REF ALT \"ANN[*].GENE\" \"ANN[*].FEATUREID\" \"ANN[*].EFFECT\" \"LOF[*].GENE\" >> {SampleName}.UG.ann.sort.txt\n",
    "!echo \"#Moderate impact mutations\" >> {SampleName}.UG.ann.sort.txt\n",
    "!cat {SampleName}.UG.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar filter \"ANN[0].IMPACT has 'MODERATE'\" | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - CHROM POS REF ALT \"ANN[*].GENE\" \"ANN[*].FEATUREID\" \"ANN[*].EFFECT\" \"LOF[*].GENE\" >> {SampleName}.UG.ann.sort.txt\n",
    "!echo \"#Low impact mutations\" >> {SampleName}.UG.ann.sort.txt\n",
    "!cat {SampleName}.UG.ann.vcf | perl {softpath}/SnpEff-5.0/snpEff/scripts/vcfEffOnePerLine.pl | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar filter \"ANN[0].IMPACT has 'LOW'\" | java -Xmx{ramG}g -jar {softpath}/SnpEff-5.0/snpEff/SnpSift.jar extractFields - CHROM POS REF ALT \"ANN[*].GENE\" \"ANN[*].FEATUREID\" \"ANN[*].EFFECT\" \"LOF[*].GENE\" >> {SampleName}.UG.ann.sort.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unified Genotyper\n",
      "#High impact mutations\n",
      "CHROM\tPOS\tREF\tALT\tANN[*].GENE\tANN[*].FEATUREID\tANN[*].EFFECT\tLOF[*].GENE\n",
      "I\t6313\tG\tT\thomt-1\tY74C9A.3.1\tstop_gained\thomt-1\n",
      "I\t59670\tCA\tC\tY48G1C.5\tY48G1C.5.1\tframeshift_variant\tY48G1C.5\n",
      "I\t218097\tA\tAT\tY48G1BL.8\tY48G1BL.8.1\tframeshift_variant\tY48G1BL.8\n",
      "I\t238696\tA\tAG\tY48G1BM.7\tY48G1BM.7.1\tframeshift_variant\tY48G1BM.7\n",
      "I\t238954\tA\tT\tY48G1BM.7\tY48G1BM.7.1\tstop_gained\tY48G1BM.7\n",
      "I\t642997\tG\tA\tY65B4A.2\tY65B4A.2c.1\tsplice_donor_variant&intron_variant\tY65B4A.2\n",
      "I\t664805\tA\tACGACATC\tY65B4A.9\tY65B4A.9a.1\tframeshift_variant\tY65B4A.9\n",
      "I\t688509\tAG\tA\tptp-5.3\tY18H1A.1.1\tframeshift_variant\tptp-5.3\n",
      "Haplotype Caller\n",
      "#High impact mutations\n",
      "CHROM\tPOS\tREF\tALT\tANN[*].GENE\tANN[*].FEATUREID\tANN[*].EFFECT\tLOF[*].GENE\n",
      "I\t6313\tG\tT\thomt-1\tY74C9A.3.1\tstop_gained\thomt-1\n",
      "I\t59670\tCA\tC\tY48G1C.5\tY48G1C.5.1\tframeshift_variant\tY48G1C.5\n",
      "I\t218097\tA\tAT\tY48G1BL.8\tY48G1BL.8.1\tframeshift_variant\tY48G1BL.8\n",
      "I\t238696\tA\tAG\tY48G1BM.7\tY48G1BM.7.1\tframeshift_variant\tY48G1BM.7\n",
      "I\t238954\tA\tT\tY48G1BM.7\tY48G1BM.7.1\tstop_gained\tY48G1BM.7\n",
      "I\t642997\tG\tA\tY65B4A.2\tY65B4A.2c.1\tsplice_donor_variant&intron_variant\tY65B4A.2\n",
      "I\t664805\tA\tACGACATC\tY65B4A.9\tY65B4A.9a.1\tframeshift_variant\tY65B4A.9\n",
      "I\t688509\tAG\tA\tptp-5.3\tY18H1A.1.1\tframeshift_variant\tptp-5.3\n"
     ]
    }
   ],
   "source": [
    "!echo \"Unified Genotyper\"\n",
    "!head CB4856.UG.ann.sort.txt\n",
    "!echo \"Haplotype Caller\"\n",
    "!head CB4856.UG.ann.sort.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Full report of SNPEff\n",
    "\n",
    "<a href=\"analysis/VariantPrediction/CB4856.UG.html\">Unified Genotyper results</a>\n",
    "\n",
    "<a href=\"analysis/VariantPrediction/CB4856.HC.html\">Haplotype Caller results</a>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Analysis completed, going back to main directory\n",
    "os.chdir(path)\n",
    "##Create directory to place copy of results and check if files are there\n",
    "!mkdir -p {path}/results\n",
    "##Move to that directory\n",
    "os.chdir( path + '/results' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make link to main results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ln: failed to create symbolic link './SRR2003569_1_fastqc.zip': File exists\n",
      "ln: failed to create symbolic link './SRR2003569_2_fastqc.zip': File exists\n",
      "ln: failed to create symbolic link './CB4856.realigned.bam': File exists\n",
      "ln: failed to create symbolic link './CB4856.g.vcf': File exists\n",
      "ln: failed to create symbolic link './CB4856.HC.vcf': File exists\n",
      "ln: failed to create symbolic link './CB4856.UG.vcf': File exists\n",
      "ln: failed to create symbolic link './CB4856.HC.ann.sort.txt': File exists\n",
      "ln: failed to create symbolic link './CB4856.HC.ann.txt': File exists\n",
      "ln: failed to create symbolic link './CB4856.HC.genes.txt': File exists\n",
      "ln: failed to create symbolic link './CB4856.UG.ann.sort.txt': File exists\n",
      "ln: failed to create symbolic link './CB4856.UG.ann.txt': File exists\n",
      "ln: failed to create symbolic link './CB4856.UG.genes.txt': File exists\n",
      "  adding: CB4856.HC.ann.sort.txt (deflated 88%)\n",
      "  adding: CB4856.HC.ann.txt (deflated 92%)\n",
      "  adding: CB4856.HC.genes.txt (deflated 86%)\n",
      "  adding: CB4856.UG.ann.sort.txt (deflated 89%)\n",
      "  adding: CB4856.UG.ann.txt (deflated 92%)\n",
      "  adding: CB4856.UG.genes.txt (deflated 85%)\n"
     ]
    }
   ],
   "source": [
    "!ln -s {path}/analysis/FastQC/*.zip .\n",
    "!ln -s {path}/analysis/Alignments/*.realigned.bam .\n",
    "!ln -s {path}/analysis/VariantCalling/*.vcf .\n",
    "!ln -s {path}/analysis/VariantPrediction/*.txt .\n",
    "!zip SnpEff.HC.zip *.HC.*.txt\n",
    "!zip SnpEff.UG.zip *.UG.*.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Results\n",
    "\n",
    "* <a href=\"results/SRR2003569_1_fastqc.zip\">FastQC R1 report</a>\n",
    "* <a href=\"results/SRR2003569_2_fastqc.zip\">FastQC R2 report</a>\n",
    "* <a href=\"results/CB4856.realigned.bam\">Realigned bam</a>\n",
    "* <a href=\"results/CB4856.HC.vcf\">Haplotype Caller VCF</a>\n",
    "* <a href=\"results/CB4856.UG.vcf\">Unified Genotyper VCF</a>\n",
    "* <a href=\"results/SnpEff.HC.zip\">Haplotype Caller SnpEff annotations</a>\n",
    "* <a href=\"results/SnpEff.UG.zip\">Unified Genotyper SnpEff annotations</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
