{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WGS analysis\n",
    "## Estimation of copy number variation\n",
    "* [Amhed Missael Vargas Velazquez](https://www.researchgate.net/profile/Amhed-Vargas-Velazquez)\n",
    "* Post-doctoral fellow, [SGB lab](https://syngenbio.kaust.edu.sa/), [KAUST](https://www.kaust.edu.sa/en)\n",
    "\n",
    "## Brief description\n",
    "This jupyter notebook describes the processing pipeline used to estimate the abundance of piRNAi fragments in *C. elegans* strains.\n",
    "\n",
    "### Methodology\n",
    "Paired end sequencing reads were mapped to an indexed reference fasta file containing the C. elegans WS235/ce11 genome sequence, synthetic piRNA fragment sequences, protein tags, and plasmid DNA sequences (bacterial backbone removed) via bwa (v.7.17-r1188) mem algorithm with default parameters. Duplicated reads were filtered out of the alignments using picard tools v2.23.6 and coverage analysis were performed using samtools (v.1.9) coverage function. Abundance of DNA sequences in extrachromosomal arrays was estimated by dividing their mean read depth over the average read depth seen in *C. elegans* chromosomes.  \n",
    "\n",
    "## Software requirements\n",
    "The core instance running this script is Python. However, the analysis are performed by multiple programs (handled by `system calls`) which have to be installed or, for portability convenience, be present in the working directory.\n",
    "* samtools\n",
    "* fastqc\n",
    "* picard\n",
    "\n",
    "## Results\n",
    "Coverage analysis showed that plasmids that have been linearilized prior injection get inserted into extrachromosomal arrays in larger numbers than those that were not (undigested)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "## Working directory and libraries import\n",
    "We import all the required Python libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load libraries\n",
    "#os to move within directories\n",
    "import os, sys\n",
    "##Set main working directory\n",
    "path = '/home/velazqam/Documents/Projects/piRNA_counts/June_2021/WGS'\n",
    "##Move to path\n",
    "os.chdir(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameter definition\n",
    "We set all relevant parameters for our notebook. By convention, parameters are uppercase, while all the \n",
    "other variables follow Python's guidelines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "minMaqQforBam=1\n",
    "minMaqQforVcf=10\n",
    "minBaseQforVcf=10\n",
    "minVarCall=10\n",
    "diffStep=10\n",
    "minDel=2\n",
    "maxSample=2\n",
    "Ncpu=28\n",
    "ramG=40"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis\n",
    "DO NOT RUN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Add reference genome\n",
    "cd Set1\n",
    "cat ../c_elegans.PRJNA13758.WS275.genomic_softmasked.fa Background.fasta Plasmids.fasta > Reference.fasta\n",
    "cd ..\n",
    "\n",
    "cd Set2\n",
    "cat ../c_elegans.PRJNA13758.WS275.genomic_softmasked.fa Background.fasta Plasmids.fasta > Reference.fasta\n",
    "cd ..\n",
    "\n",
    "cd Set3\n",
    "cat ../c_elegans.PRJNA13758.WS275.genomic_softmasked.fa Plasmids.fasta > Reference.fasta\n",
    "cd ..\n",
    "\n",
    "sami=/home/velazqam/Documents/Projects/piRNA_counts/June_2021/WGS/scratch/samtools-1.12/bin/bin/samtools\n",
    "\n",
    "##Start with simplest one\n",
    "cd Set3\n",
    "for lib in `ls -d S*`; do\n",
    "echo ${lib}\n",
    "cd ${lib}\n",
    "cp ../Reference.fasta .\n",
    "samtools faidx Reference.fasta\n",
    "bwa index Reference.fasta\n",
    "java -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar CreateSequenceDictionary R=Reference.fasta O=Reference.dict\n",
    "\n",
    "#Perform fastqc\n",
    "fastqc -t $Ncpu ${lib}*.gz;\n",
    "\n",
    "##Perform mapping\n",
    "bwa mem -t $Ncpu -aM Reference.fasta ${lib}*.gz | samtools view -buS - |samtools sort - -o ${lib}_rawmap.bam\n",
    "\n",
    "#Samtools sort\n",
    "samtools sort -@ $Ncpu ${lib}_rawmap.bam > ${lib}.sort.bam\n",
    "#rm ${lib}_rawmap.bam\n",
    "\n",
    "#Samtools index\n",
    "samtools index ${lib}.sort.bam\n",
    "\n",
    "##Remove minimal quality for Bam\n",
    "samtools view -@ $Ncpu -q $minMaqQforBam -bh ${lib}.sort.bam > ${lib}.UM.bam\n",
    "#rm ${lib}.sort.bam ${lib}.sort.bam.bai\n",
    "\n",
    "#Picard add replace groups ## Explain its importance for combining multiple libraries\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar AddOrReplaceReadGroups I=${lib}.UM.bam O=${lib}.RG.bam RGID=${lib} RGLB=LB RGPL=illumina RGPU=PU RGSM=${lib}\n",
    "#rm ${lib}.UM.bam\n",
    "\n",
    "#Picard remove duplicates\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar MarkDuplicates I=${lib}.RG.bam O=${lib}.deDup.bam M=${lib}.dedupMetrics REMOVE_DUPLICATES=true\n",
    "#rm ${lib}.RG.bam ${lib}.dedupMetrics\n",
    "\n",
    "##Samtools\n",
    "samtools index ${lib}.deDup.bam\n",
    "\n",
    "${sami} coverage ${lib}.deDup.bam > ${lib}.coverage.tsv\n",
    "\n",
    "cd ..;\n",
    "\n",
    "done\n",
    "\n",
    "cd ..\n",
    "\n",
    "##Start with simplest one\n",
    "cd Set2\n",
    "for lib in `ls -d S*`; do\n",
    "echo ${lib}\n",
    "cd ${lib}\n",
    "cp ../Reference.fasta .\n",
    "samtools faidx Reference.fasta\n",
    "bwa index Reference.fasta\n",
    "java -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar CreateSequenceDictionary R=Reference.fasta O=Reference.dict\n",
    "\n",
    "#Perform fastqc\n",
    "fastqc -t $Ncpu ${lib}*.gz;\n",
    "\n",
    "##Perform mapping\n",
    "bwa mem -t $Ncpu -aM Reference.fasta ${lib}*.gz | samtools view -buS - |samtools sort - -o ${lib}_rawmap.bam\n",
    "\n",
    "#Samtools sort\n",
    "samtools sort -@ $Ncpu ${lib}_rawmap.bam > ${lib}.sort.bam\n",
    "#rm ${lib}_rawmap.bam\n",
    "\n",
    "#Samtools index\n",
    "samtools index ${lib}.sort.bam\n",
    "\n",
    "##Remove minimal quality for Bam\n",
    "samtools view -@ $Ncpu -q $minMaqQforBam -bh ${lib}.sort.bam > ${lib}.UM.bam\n",
    "#rm ${lib}.sort.bam ${lib}.sort.bam.bai\n",
    "\n",
    "#Picard add replace groups ## Explain its importance for combining multiple libraries\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar AddOrReplaceReadGroups I=${lib}.UM.bam O=${lib}.RG.bam RGID=${lib} RGLB=LB RGPL=illumina RGPU=PU RGSM=${lib}\n",
    "#rm ${lib}.UM.bam\n",
    "\n",
    "#Picard remove duplicates\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar MarkDuplicates I=${lib}.RG.bam O=${lib}.deDup.bam M=${lib}.dedupMetrics REMOVE_DUPLICATES=true\n",
    "#rm ${lib}.RG.bam ${lib}.dedupMetrics\n",
    "\n",
    "##Samtools\n",
    "samtools index ${lib}.deDup.bam\n",
    "\n",
    "${sami} coverage ${lib}.deDup.bam > ${lib}.coverage.tsv\n",
    "\n",
    "##GATK Indel realigner\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T RealignerTargetCreator -nt $Ncpu -R Reference.fasta -I ${lib}.deDup.bam -o ${lib}.forIndelRealigner.intervals\n",
    "\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T IndelRealigner -R Reference.fasta -I ${lib}.deDup.bam -targetIntervals ${lib}.forIndelRealigner.intervals -o ${lib}.realigned.bam\n",
    "\n",
    "#rm ${lib}.deDup.bam ${lib}.deDup.bam.bai ${lib}.forIndelRealigner.intervals\n",
    "\n",
    "#Unified genotyper\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T UnifiedGenotyper -R Reference.fasta -nt $Ncpu -l INFO -glm BOTH -I ${lib}.realigned.bam -o ${lib}.UG.vcf -mbq $minBaseQforVcf -stand_call_conf $minVarCall \n",
    "\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T HaplotypeCaller --emitRefConfidence GVCF -R Reference.fasta -l INFO -I ${lib}.realigned.bam -o ${lib}.g.vcf -mbq ${minBaseQforVcf} -mmq ${minMaqQforVcf} -stand_call_conf ${minVarCall}\n",
    "\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T GenotypeGVCFs -R Reference.fasta -nt ${Ncpu} -l INFO -V ${lib}.g.vcf -dt none -o ${lib}.HC.vcf\n",
    "\n",
    "cd ..;\n",
    "\n",
    "done\n",
    "\n",
    "cd ..\n",
    "\n",
    "\n",
    "##Start with simplest one\n",
    "cd Set1\n",
    "for lib in `ls -d *_*`; do\n",
    "echo ${lib}\n",
    "cd ${lib}\n",
    "cp ../Reference.fasta .\n",
    "samtools faidx Reference.fasta\n",
    "bwa index Reference.fasta\n",
    "java -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar CreateSequenceDictionary R=Reference.fasta O=Reference.dict\n",
    "\n",
    "#Perform fastqc\n",
    "fastqc -t $Ncpu ${lib}*.gz;\n",
    "\n",
    "##Perform mapping\n",
    "bwa mem -t $Ncpu -aM Reference.fasta ${lib}*.gz | samtools view -buS - |samtools sort - -o ${lib}_rawmap.bam\n",
    "\n",
    "#Samtools sort\n",
    "samtools sort -@ $Ncpu ${lib}_rawmap.bam > ${lib}.sort.bam\n",
    "#rm ${lib}_rawmap.bam\n",
    "\n",
    "#Samtools index\n",
    "samtools index ${lib}.sort.bam\n",
    "\n",
    "##Remove minimal quality for Bam\n",
    "samtools view -@ $Ncpu -q $minMaqQforBam -bh ${lib}.sort.bam > ${lib}.UM.bam\n",
    "#rm ${lib}.sort.bam ${lib}.sort.bam.bai\n",
    "\n",
    "#Picard add replace groups ## Explain its importance for combining multiple libraries\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar AddOrReplaceReadGroups I=${lib}.UM.bam O=${lib}.RG.bam RGID=${lib} RGLB=LB RGPL=illumina RGPU=PU RGSM=${lib}\n",
    "#rm ${lib}.UM.bam\n",
    "\n",
    "#Picard remove duplicates\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/picard2.23.6/picard.jar MarkDuplicates I=${lib}.RG.bam O=${lib}.deDup.bam M=${lib}.dedupMetrics REMOVE_DUPLICATES=true\n",
    "#rm ${lib}.RG.bam ${lib}.dedupMetrics\n",
    "\n",
    "##Samtools\n",
    "samtools index ${lib}.deDup.bam\n",
    "\n",
    "${sami} coverage ${lib}.deDup.bam > ${lib}.coverage.tsv\n",
    "\n",
    "##GATK Indel realigner\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T RealignerTargetCreator -nt $Ncpu -R Reference.fasta -I ${lib}.deDup.bam -o ${lib}.forIndelRealigner.intervals\n",
    "\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T IndelRealigner -R Reference.fasta -I ${lib}.deDup.bam -targetIntervals ${lib}.forIndelRealigner.intervals -o ${lib}.realigned.bam\n",
    "\n",
    "#rm ${lib}.deDup.bam ${lib}.deDup.bam.bai ${lib}.forIndelRealigner.intervals\n",
    "\n",
    "#Unified genotyper\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T UnifiedGenotyper -R Reference.fasta -nt $Ncpu -l INFO -glm BOTH -I ${lib}.realigned.bam -o ${lib}.UG.vcf -mbq $minBaseQforVcf -stand_call_conf $minVarCall \n",
    "\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T HaplotypeCaller --emitRefConfidence GVCF -R Reference.fasta -l INFO -I ${lib}.realigned.bam -o ${lib}.g.vcf -mbq ${minBaseQforVcf} -mmq ${minMaqQforVcf} -stand_call_conf ${minVarCall}\n",
    "\n",
    "java -Xmx${ramG}g -jar /home/velazqam/Downloads/Sonia_WGS/Software/gatk-3.8.1.0/GenomeAnalysisTK.jar -T GenotypeGVCFs -R Reference.fasta -nt ${Ncpu} -l INFO -V ${lib}.g.vcf -dt none -o ${lib}.HC.vcf\n",
    "\n",
    "cd ..;\n",
    "\n",
    "done\n",
    "\n",
    "cd ..\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References\n",
    "* Li, H. & Durbin, R. Fast and accurate short read alignment with Burrows–Wheeler transform. Bioinformatics 25, 1754–1760 (2009).\n",
    "* Li, H. et al. The Sequence Alignment/Map format and SAMtools. Bioinformatics 25, 2078–2079 (2009).\n",
    "* “Picard Tools.” Broad Institute, GitHub repository. http://broadinstitute.github.io/picard/."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
